<!DOCTYPE html>
<html lang="en-us"><head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="generator" content="Hugo 0.62.2" />
  
  <title>Roman Hillje - Data Visualization &amp; Bioinformatics</title>
  
  
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.0.13/css/all.css" integrity="sha384-DNOHZ68U8hZfKXOrtjWvjxusGo9WQnrNx2sqG0tfsghAvtVlRW3tvkXWZh58N9jp" crossorigin="anonymous">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Righteous%7CMerriweather:300,300i,400,400i,700,700i">
  <link rel="stylesheet" href="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.css"/>
  <link rel="stylesheet" href="/css/medium.css">
</head>
<body><nav class="navbar navbar-expand-lg navbar-light bg-white mediumnavigation nav-down navbar-static-top">
  <div class="container pr-0">

    
    <a class="navbar-brand" href="http://romanhaa.github.io/">
      
      <span style="font-family:Righteous;">Roman Hillje - Data Visualization &amp; Bioinformatics</span>
      
    </a>

    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarMediumish" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    
    <div class="collapse navbar-collapse" id="navbarMediumish">
      
      <ul class="navbar-nav ml-auto">
        
        <li class="nav-item ">
          <a class="nav-link" href="/">About Me</a>
        </li>
        
        <li class="nav-item ">
          <a class="nav-link" href="/projects">Projects</a>
        </li>
        
        <li class="nav-item ">
          <a class="nav-link" href="/plots">Plot Gallery</a>
        </li>
        
        <li class="nav-item ">
          <a class="nav-link" href="/blog">Blog</a>
        </li>
        
      </ul>
    </div>

  </div>
</nav>
<div class="site-content">
      <div class="container">
  <link rel="stylesheet" href="/css/numbered_headers.css">
  <div class="main-content">
    
    <div class="container">
      <div class="row">
        
        <div class="col-md-12 flex-first flex-md-unordered">
          <div class="mainheading">
            
            <h1 class="posttitle">Animating the dynamics of training a Poincaré map</h1>
            <div class="article-post">
              Mar 1, 2020
            </div>
          </div>
          
          
          
          <body>
            <div class="article-post">
              <h2 id="preface">Preface</h2>
<p>Last Friday, I came across an interesting method called <a href="https://www.biorxiv.org/content/10.1101/689547v2">&ldquo;Poincaré map&rdquo;</a> to visualize the relationship of cells in single cell RNA-seq data using hyperbolic space.
Check <a href="https://dmishin.github.io/hyperbolic-ca-simulator/index.html">this page</a> for an intuitive representation of what hyperbolic space is.
During the weekend, I had the chance to experiment with this method to the data set I used in the <a href="http://romanhaa.github.io/projects/scrnaseq_workflow/">scRNA-seq workflow</a>.
This article is not about finding the right parameters, even if it might help doing so.
Instead, it is about (1) finally getting a chance to make animated plots, and (2) visualize what happens during the training process of the Poincaré map.</p>
<p>Links:</p>
<ul>
<li>Preprint on bioRxiv: <a href="https://www.biorxiv.org/content/10.1101/689547v2">Poincaré Maps for Analyzing Complex Hierarchies in Single-Cell Data</a></li>
<li>GitHub repository: <a href="https://github.com/facebookresearch/PoincareMaps">facebookresearch/PoincareMaps</a></li>
</ul>
<p>Let's go.</p>
<h2 id="export-log-normalized-transcript-counts">Export log-normalized transcript counts</h2>
<p>First, we will load our data and identify genes with a mean expression above a visually determined and quite arbitrary threshold (I'm not sure this is the right way to do it but I wanted to reduce computation time).
We remove almost half of the genes, having 8,886 genes left.</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-r" data-lang="r"><span style="color:#900;font-weight:bold">library</span>(tidyverse)
<span style="color:#900;font-weight:bold">library</span>(Seurat)

seurat <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">readRDS</span>(<span style="color:#d14">&#39;</span><span style="color:#d14">data/seurat.rds&#39;</span>)

average_expression_per_gene <span style="color:#000;font-weight:bold">&lt;-</span> seurat<span style="color:#000;font-weight:bold">@</span>assays<span style="color:#000;font-weight:bold">$</span>SCT<span style="color:#000;font-weight:bold">@</span>data <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowMeans</span>()

p <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(value <span style="color:#000;font-weight:bold">=</span> average_expression_per_gene) <span style="color:#000;font-weight:bold">%&gt;%</span>
  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(value)) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">geom_density</span>(fill <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">grey&#39;</span>, alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.8</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">geom_vline</span>(xintercept <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.01</span>, color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">red&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">scale_x_log10</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">Mean log-normalized expression&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">scale_y_continuous</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">Density&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">theme_bw</span>()
<span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;</span><span style="color:#d14">mean_log_normalized_expression.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">7</span>)

<span style="color:#900;font-weight:bold">length</span>(average_expression_per_gene)
<span style="color:#998;font-style:italic"># 15907</span>

<span style="color:#900;font-weight:bold">length</span>(<span style="color:#900;font-weight:bold">which</span>(average_expression_per_gene <span style="color:#000;font-weight:bold">&gt;=</span> <span style="color:#099">0.01</span>))
<span style="color:#998;font-style:italic"># 8886</span>
</code></pre></div>

<img src="/images/blog/dynamics_of_training_a_poincare_map/mean_log_normalized_expression.png" style="max-width: 600px; display: block; margin-left: auto; margin-right: auto;" />


<p>Then, we write the filtered transcript counts in a transposed format to a CSV file, adding the cell type identified using <code>SingleR</code> as a label.</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-r" data-lang="r">seurat<span style="color:#000;font-weight:bold">@</span>assays<span style="color:#000;font-weight:bold">$</span>SCT<span style="color:#000;font-weight:bold">@</span>data<span style="color:#900;font-weight:bold">[which</span>(average_expression_per_gene <span style="color:#000;font-weight:bold">&gt;=</span> <span style="color:#099">0.01</span>),] <span style="color:#000;font-weight:bold">%&gt;%</span>
<span style="color:#900;font-weight:bold">as.matrix</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
<span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
<span style="color:#900;font-weight:bold">as_tibble</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
<span style="color:#900;font-weight:bold">mutate</span>(label <span style="color:#000;font-weight:bold">=</span> seurat<span style="color:#000;font-weight:bold">@</span>meta.data<span style="color:#000;font-weight:bold">$</span>cell_type_singler_blueprintencode_main) <span style="color:#000;font-weight:bold">%&gt;%</span>
<span style="color:#900;font-weight:bold">write_csv</span>(<span style="color:#d14">&#39;</span><span style="color:#d14">data/SCT_data.csv&#39;</span>)
</code></pre></div><h2 id="adjust-code-of-poincaré-tool">Adjust code of Poincaré tool</h2>
<p>Before actually generating any Poincaré map, we have to download the tool from GitHub and a few small adjustments to the code.
The tool allows to see intermediate maps (and cell coordinates) through the <code>--debugplot</code> parameter where we specify how often we want to get the intermediate data.
If we set it to <code>10</code> (which we will do later), the intermediate data is generated every 10 epochs.
The adjustments in the code are necessary because the intermediate data is overwritten every time they are re-generated.
In the <code>train.py</code> file, in the definition of the <code>train</code> function, we create a new variable <code>temp_fout</code>, which will be the file name of the temporary data, and add the current epoch count to the file name.
Then, we use that temporary name for the intermediate map (in PDF format) and cell coordinates (CSV format).</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-py" data-lang="py"><span style="color:#998;font-style:italic"># ...</span>
        <span style="color:#000;font-weight:bold">if</span> args<span style="color:#000;font-weight:bold">.</span>debugplot:
            <span style="color:#000;font-weight:bold">if</span> (epoch <span style="color:#000;font-weight:bold">%</span> args<span style="color:#000;font-weight:bold">.</span>debugplot) <span style="color:#000;font-weight:bold">==</span> <span style="color:#099">0</span>:
                d <span style="color:#000;font-weight:bold">=</span> model<span style="color:#000;font-weight:bold">.</span>lt<span style="color:#000;font-weight:bold">.</span>weight<span style="color:#000;font-weight:bold">.</span>cpu()<span style="color:#000;font-weight:bold">.</span>detach()<span style="color:#000;font-weight:bold">.</span>numpy()
                titlename <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14"></span><span style="color:#d14">&#39;</span><span style="color:#d14">epoch: {:d}, loss: {:.3e}</span><span style="color:#d14">&#39;</span><span style="color:#000;font-weight:bold">.</span>format(
                    epoch, np<span style="color:#000;font-weight:bold">.</span>mean(epoch_loss))
                temp_fout <span style="color:#000;font-weight:bold">=</span> fout <span style="color:#000;font-weight:bold">+</span> <span style="color:#d14"></span><span style="color:#d14">&#34;</span><span style="color:#d14">_epoch=</span><span style="color:#d14">&#34;</span> <span style="color:#000;font-weight:bold">+</span> <span style="color:#0086b3">str</span>(epoch)

                <span style="color:#000;font-weight:bold">if</span> epoch <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">5</span>:
                    plotPoincareDisc(np<span style="color:#000;font-weight:bold">.</span>transpose(d), labels, temp_fout, titlename, color_dict<span style="color:#000;font-weight:bold">=</span>color_dict)
                    np<span style="color:#000;font-weight:bold">.</span>savetxt(temp_fout <span style="color:#000;font-weight:bold">+</span> <span style="color:#d14"></span><span style="color:#d14">&#39;</span><span style="color:#d14">.csv</span><span style="color:#d14">&#39;</span>, d, delimiter<span style="color:#000;font-weight:bold">=</span><span style="color:#d14"></span><span style="color:#d14">&#34;</span><span style="color:#d14">,</span><span style="color:#d14">&#34;</span>)
<span style="color:#998;font-style:italic"># ...</span>
</code></pre></div><p>I think this is a quite ugly approach but it gets the job done for now.
If the developers think this is useful, I would think about a way to store intermediate results without creating many separate files.</p>
<h2 id="learn-poincaré-map">Learn Poincaré map</h2>
<p>Now, we can start training the Poincaré map using the same parameters as documented for some of the examples in the GitHub repository.
By default, this process will run for 5,000 epochs (<code>--epochs</code>) or until epsilon (loss) goes below <code>0.0001</code> (<code>--earlystop</code>).
I did my early tests on my private laptop which doesn't have a GPU, which means I had to set <code>--cuda</code> to <code>0</code>.
I terminated the command below after a about an hour because it had only calculated 290 epochs.
Further tests will have to be run on some proper hardware, but for the purpose of visualizing the process we anyway collected some interesting data.</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-sh" data-lang="sh">python3 main.py <span style="color:#d14">\
</span><span style="color:#d14"></span>--dset SCT_data <span style="color:#d14">\
</span><span style="color:#d14"></span>--path data/ <span style="color:#d14">\
</span><span style="color:#d14"></span>--batchsize -1 <span style="color:#d14">\
</span><span style="color:#d14"></span>--cuda <span style="color:#099">0</span> <span style="color:#d14">\
</span><span style="color:#d14"></span>--knn <span style="color:#099">15</span> <span style="color:#d14">\
</span><span style="color:#d14"></span>--gamma 2.0 <span style="color:#d14">\
</span><span style="color:#d14"></span>--sigma 1.0 <span style="color:#d14">\
</span><span style="color:#d14"></span>--pca <span style="color:#099">20</span> <span style="color:#d14">\
</span><span style="color:#d14"></span>--labels <span style="color:#099">1</span> <span style="color:#d14">\
</span><span style="color:#d14"></span>--mode features <span style="color:#d14">\
</span><span style="color:#d14"></span>--debugplot <span style="color:#099">10</span>
</code></pre></div><p>The plot below is generated by the Poincaré tool and shows the loss curve during the training process.
If I interpret it correctly, the top value indicates that we were actually very close to the <code>0.0001</code> epsilon that would've stopped the analysis.
But maybe I'm wrong, there isn't too much information.</p>


<img src="/images/blog/dynamics_of_training_a_poincare_map/loss_curve.pdf" style="width: 50%; display: block; margin-left: auto; margin-right: auto;" />


<p>Anyway, all the intermediate data was put into the <code>results</code> folder of the current working directory.</p>
<h2 id="prepare-r-session">Prepare R session</h2>
<p>Here, we launch an R session, load libraries, load our Seurat object, and define some custom colors.</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-r" data-lang="r"><span style="color:#900;font-weight:bold">library</span>(tidyverse)
<span style="color:#900;font-weight:bold">library</span>(Seurat)
<span style="color:#900;font-weight:bold">library</span>(plotly)

seurat <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">readRDS</span>(<span style="color:#d14">&#39;</span><span style="color:#d14">data/seurat.rds&#39;</span>)

colors_dutch <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">c</span>(
  <span style="color:#d14">&#39;</span><span style="color:#d14">#FFC312&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#C4E538&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#12CBC4&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#FDA7DF&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#ED4C67&#39;</span>,
  <span style="color:#d14">&#39;</span><span style="color:#d14">#F79F1F&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#A3CB38&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#1289A7&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#D980FA&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#B53471&#39;</span>,
  <span style="color:#d14">&#39;</span><span style="color:#d14">#EE5A24&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#009432&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#0652DD&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#9980FA&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#833471&#39;</span>,
  <span style="color:#d14">&#39;</span><span style="color:#d14">#EA2027&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#006266&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#1B1464&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#5758BB&#39;</span>,<span style="color:#d14">&#39;</span><span style="color:#d14">#6F1E51&#39;</span>
)
</code></pre></div><h2 id="load-data">Load data</h2>
<p>Now, we load the cell coordinates from all available intermediate results.
While doing this, we also add an ID and the cell type to every cell.</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-r" data-lang="r">final_data <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
  X1 <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>(),
  X2 <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>(),
  cell_id <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>(),
  epoch <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>(),
  label <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">character</span>()
)

<span style="color:#900;font-weight:bold">for </span>( i in <span style="color:#900;font-weight:bold">list.files</span>(<span style="color:#d14">&#39;</span><span style="color:#d14">~/Research/PoincareMaps/results&#39;</span>, pattern <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">seed0_epoch=[0-9]{2,4}.csv&#39;</span>, full.name <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">TRUE</span>) ) {
  epoch <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">gsub</span>(<span style="color:#900;font-weight:bold">basename</span>(i), pattern <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">SCT_data_PM15sigma=1.00gamma=2.00minkowskipca=20_seed0_epoch=&#39;</span>, replacement <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">&#39;</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
    <span style="color:#900;font-weight:bold">gsub</span>(pattern <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">\\.csv&#39;</span>, replacement <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">&#39;</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
    <span style="color:#900;font-weight:bold">as.integer</span>()
  temp_data <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">read_csv</span>(i, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>, col_types <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">cols</span>()) <span style="color:#000;font-weight:bold">%&gt;%</span>
    <span style="color:#900;font-weight:bold">mutate</span>(
      cell_id <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">row_number</span>(),
      epoch <span style="color:#000;font-weight:bold">=</span> epoch,
      label <span style="color:#000;font-weight:bold">=</span> seurat<span style="color:#000;font-weight:bold">@</span>meta.data<span style="color:#000;font-weight:bold">$</span>cell_type_singler_blueprintencode_main
    )
  final_data <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">bind_rows</span>(final_data, temp_data)
}

final_data <span style="color:#000;font-weight:bold">&lt;-</span> final_data <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">arrange</span>(epoch, cell_id)

<span style="color:#900;font-weight:bold">glimpse</span>(final_data)
<span style="color:#998;font-style:italic"># Observations: 165,213</span>
<span style="color:#998;font-style:italic"># Variables: 5</span>
<span style="color:#998;font-style:italic"># $ X1      &lt;dbl&gt; 0.004369234, 0.132621959, 0.127156094, -0.051962741, -0.22114…</span>
<span style="color:#998;font-style:italic"># $ X2      &lt;dbl&gt; 0.03714102, 0.13966362, 0.13251643, -0.06476162, -0.18119001,…</span>
<span style="color:#998;font-style:italic"># $ cell_id &lt;dbl&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18…</span>
<span style="color:#998;font-style:italic"># $ epoch   &lt;dbl&gt; 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 1…</span>
<span style="color:#998;font-style:italic"># $ label   &lt;chr&gt; &#34;HSC&#34;, &#34;CD8+ T-cells&#34;, &#34;CD8+ T-cells&#34;, &#34;Erythrocytes&#34;, &#34;Eryth…</span>
</code></pre></div><p>Also, we create a named vector that holds the color assignment for each cell type.</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-r" data-lang="r">colors_here <span style="color:#000;font-weight:bold">&lt;-</span> colors_dutch<span style="color:#000;font-weight:bold">$</span>discrete[1<span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">length</span>(<span style="color:#900;font-weight:bold">unique</span>(final_data<span style="color:#000;font-weight:bold">$</span>label))]
<span style="color:#900;font-weight:bold">names</span>(colors_here) <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">sort</span>(<span style="color:#900;font-weight:bold">unique</span>(final_data<span style="color:#000;font-weight:bold">$</span>label))
</code></pre></div><h2 id="make-static-plot">Make static plot</h2>
<p>To get a sense of the data, I first created a faceted static plot using <code>ggplot2</code> for the first 1,000 cell ID at every epoch.</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-r" data-lang="r">data_to_plot <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">filter</span>(final_data, cell_id <span style="color:#000;font-weight:bold">%in%</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1000</span>))

<span style="color:#900;font-weight:bold">glimpse</span>(data_to_plot)
<span style="color:#998;font-style:italic"># Observations: 29,000</span>
<span style="color:#998;font-style:italic"># Variables: 5</span>
<span style="color:#998;font-style:italic"># $ X1      &lt;dbl&gt; 0.004369234, 0.132621959, 0.127156094, -0.051962741, -0.22114…</span>
<span style="color:#998;font-style:italic"># $ X2      &lt;dbl&gt; 0.03714102, 0.13966362, 0.13251643, -0.06476162, -0.18119001,…</span>
<span style="color:#998;font-style:italic"># $ cell_id &lt;dbl&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18…</span>
<span style="color:#998;font-style:italic"># $ epoch   &lt;dbl&gt; 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 1…</span>
<span style="color:#998;font-style:italic"># $ label   &lt;chr&gt; &#34;HSC&#34;, &#34;CD8+ T-cells&#34;, &#34;CD8+ T-cells&#34;, &#34;Erythrocytes&#34;, &#34;Eryth…</span>

p <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">ggplot</span>(data_to_plot, <span style="color:#900;font-weight:bold">aes</span>(X1, X2)) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">geom_point</span>(<span style="color:#900;font-weight:bold">aes</span>(color <span style="color:#000;font-weight:bold">=</span> label), size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.5</span>, show.legend <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">TRUE</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">geom_circle</span>(<span style="color:#900;font-weight:bold">aes</span>(x0 <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, y0 <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, r <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>), inherit.aes <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">scale_color_manual</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">Cell type&#39;</span>, values <span style="color:#000;font-weight:bold">=</span> custom_colors<span style="color:#000;font-weight:bold">$</span>discrete) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">coord_fixed</span>() <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">theme_void</span>() <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">facet_wrap</span>(<span style="color:#000;font-weight:bold">~</span>epoch, ncol <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">8</span>)

<span style="color:#900;font-weight:bold">system.time</span>({ <span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;</span><span style="color:#d14">poincare_1000_cells_split.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">18</span>) })
<span style="color:#998;font-style:italic">#    user  system elapsed</span>
<span style="color:#998;font-style:italic"># 133.342  37.621 186.061</span>
</code></pre></div>
<figure >
  
    <a href="/images/blog/dynamics_of_training_a_poincare_map/poincare_1000_cells_split.png" target="_blank">
      <img src="/images/blog/dynamics_of_training_a_poincare_map/poincare_1000_cells_split.jpg"  style="width: 100%;" />
    </a>
  
  
</figure>


<p>As documented in the code block, this plot already took 3 minutes to generate.
I found this unusually long but I don't know of a way to know if it's just a feeling or a correct assessment.</p>
<h2 id="create-animation-with-gganimate">Create animation with <code>gganimate</code></h2>
<p>My initial idea was to create the animation using <a href="https://github.com/thomasp85/gganimate"><code>gganimate</code></a>.
However, after a few attempts, I realized that it takes way too long for it to render.
I was able to make on with the first 100 cell IDs (it took 5 minutes), but even with just 1,000 cells the process crashed due to a lack of memory after about 1 hour.</p>
<p><strong>Note:</strong> The full data set contains ~5,700 cells.</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-r" data-lang="r">data_to_plot <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">filter</span>(final_data, cell_id <span style="color:#000;font-weight:bold">%in%</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">100</span>))

<span style="color:#900;font-weight:bold">glimpse</span>(data_to_plot)
<span style="color:#998;font-style:italic"># Observations: 2,900</span>
<span style="color:#998;font-style:italic"># Variables: 5</span>
<span style="color:#998;font-style:italic"># $ X1      &lt;dbl&gt; 0.004369234, 0.132621959, 0.127156094, -0.051962741, -0.22114…</span>
<span style="color:#998;font-style:italic"># $ X2      &lt;dbl&gt; 0.03714102, 0.13966362, 0.13251643, -0.06476162, -0.18119001,…</span>
<span style="color:#998;font-style:italic"># $ cell_id &lt;dbl&gt; 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18…</span>
<span style="color:#998;font-style:italic"># $ epoch   &lt;dbl&gt; 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 1…</span>
<span style="color:#998;font-style:italic"># $ label   &lt;chr&gt; &#34;HSC&#34;, &#34;CD8+ T-cells&#34;, &#34;CD8+ T-cells&#34;, &#34;Erythrocytes&#34;, &#34;Eryth…</span>

p <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">ggplot</span>(data_to_plot, <span style="color:#900;font-weight:bold">aes</span>(X1, X2)) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">geom_point</span>(<span style="color:#900;font-weight:bold">aes</span>(color <span style="color:#000;font-weight:bold">=</span> label), alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.7</span>, show.legend <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">TRUE</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">geom_circle</span>(<span style="color:#900;font-weight:bold">aes</span>(x0 <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, y0 <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, r <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>), inherit.aes <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">scale_color_manual</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">Cell type&#39;</span>, values <span style="color:#000;font-weight:bold">=</span> custom_colors<span style="color:#000;font-weight:bold">$</span>discrete) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">coord_fixed</span>() <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">theme_void</span>() <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">labs</span>(title <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">Epoch: {format(frame_time, digits = 0)}&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">transition_time</span>(epoch) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">shadow_wake</span>(wake_length <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.1</span>, alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">+</span>
  <span style="color:#900;font-weight:bold">ease_aes</span>(<span style="color:#d14">&#39;</span><span style="color:#d14">linear&#39;</span>)

<span style="color:#900;font-weight:bold">system.time</span>({
  animation <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">animate</span>(p)
  <span style="color:#900;font-weight:bold">anim_save</span>(<span style="color:#d14">&#39;</span><span style="color:#d14">poincare_100_cells_with_tail.gif&#39;</span>, animation)
})
<span style="color:#998;font-style:italic">#    user  system elapsed</span>
<span style="color:#998;font-style:italic"># 280.944  46.455 332.494</span>
</code></pre></div>

<img src="/images/blog/dynamics_of_training_a_poincare_map/poincare_100_cells_with_tail.gif" style="width: 480px; display: block; margin-left: auto; margin-right: auto;" />


<p>The performance bottleneck is quite a bummer because I like that <code>gganimate</code> can output a GIF file which can be shared in multiple ways.
Also the added tail is quite neat, even though perhaps not very useful with all (~5,000) cells.</p>
<h2 id="create-animation-with-plotly">Create animation with <code>plotly</code></h2>
<p>After unsuccessfully trying to get around the performance problems in <code>ggplot2</code>, I started looking for alternatives and ended up with the <code>plotly</code> framework.
I worked with <code>plotly</code> quite often and thought I knew it relatively well.
Yet, I was not aware of its capability to generate animations.
And it's actually also quite easy to set up, using a similar syntax as <code>gganimate</code> (or maybe it's the other way around, who knows).</p>
<div class="highlight"><pre style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2"><code class="language-r" data-lang="r">fig <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">plot_ly</span>(
    final_data,
    type <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">scatter&#39;</span>,
    mode <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">markers&#39;</span>,
    x <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span>X1,
    y <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span>X2,
    color <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span>label,
    colors <span style="color:#000;font-weight:bold">=</span> colors_here,
    frame <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span>epoch,
    marker <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">list</span>(size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">8</span>),
    hoverinfo <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">text&#39;</span>,
    text <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span><span style="color:#900;font-weight:bold">paste0</span>(final_data<span style="color:#000;font-weight:bold">$</span>label, <span style="color:#d14">&#39;</span><span style="color:#d14">&lt;br&gt;Cell ID: &#39;</span>, <span style="color:#900;font-weight:bold">format</span>(final_data<span style="color:#000;font-weight:bold">$</span>cell_id, big.mark <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">,&#39;</span>))
  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
  <span style="color:#900;font-weight:bold">add_markers</span>(
    data <span style="color:#000;font-weight:bold">=</span> final_data <span style="color:#000;font-weight:bold">%&gt;%</span>
      dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(epoch) <span style="color:#000;font-weight:bold">%&gt;%</span>
      <span style="color:#900;font-weight:bold">distinct</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
      <span style="color:#900;font-weight:bold">mutate</span>(x <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, y <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, label <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">center&#39;</span>),
    x <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span>x,
    y <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span>y,
    frame <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span>epoch,
    marker <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">list</span>(
      size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>,
      line <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">list</span>(
        color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">black&#39;</span>,
        width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">2</span>
      ),
      symbol <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">x-thin&#39;</span>
    ),
    hoverinfo <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">skip&#39;</span>,
    text <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">~</span>label,
    showlegend <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>
  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
  <span style="color:#900;font-weight:bold">layout</span>(
    title <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">Dynamics of training a Poincaré map&#39;</span>,
    shapes <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">list</span>(
      <span style="color:#900;font-weight:bold">list</span>(
        type <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">circle&#39;</span>,
        xref <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">x&#39;</span>,
        x0 <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">-1</span>,
        x1 <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>,
        yref <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">y&#39;</span>,
        y0 <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">-1</span>,
        y1 <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>,
        fillcolor <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">NA</span>,
        line <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">list</span>(color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">black&#39;</span>),
        opacity <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>
      )
    ),
    xaxis <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">list</span>(
      range <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">-1</span>, <span style="color:#099">1</span>),
      showgrid <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>,
      showticklabels <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>,
      zeroline <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>,
      title <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">&#39;</span>
    ),
    yaxis <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">list</span>(
      range <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">-1</span>, <span style="color:#099">1</span>),
      showgrid <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>,
      showticklabels <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>,
      zeroline <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>,
      title <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">&#39;</span>,
      scaleanchor <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;</span><span style="color:#d14">x&#39;</span>
    )
  )

htmlwidgets<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">saveWidget</span>(fig, <span style="color:#d14">&#39;</span><span style="color:#d14">poincare_map.html&#39;</span>)
</code></pre></div><p>Below you find a link to the interactive result, including all cells.
Just be warned, it is <strong>16 MB</strong> in size.</p>


<a href="/images/blog/dynamics_of_training_a_poincare_map/poincare_map.html"><img src="/images/blog/dynamics_of_training_a_poincare_map/poincare_map.png" style="width: 80%; display: block; margin-left: auto; margin-right: auto;" /></a>


<h2 id="conclusion">Conclusion</h2>
<p>I must say that I'm quite impressed overall.</p>
<p>Firstly, the <code>plotly</code> framework can create amazing animations for decently large data sets, which apparently make <code>ggplot2</code> struggle.</p>
<p>Secondly, even though we used non-adjusted parameters and didn't let the training process finish (it might've finished early but I'm not sure), the resulting map has the hematopoietic stem cells (HSC) at its center.
While these results are <strong>very</strong> preliminary, they do make biological sense since these cells should give rise to the other cell types shown in the map.
This is an exciting result and makes me look forward to additional tests (hopefully with access to a GPU).</p>

            </div>
          </body>
        </div>
      </div>
    </div>
  </div>
  <script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
  <script src="/js/fold_code_boxes.js"></script>
  <script src="/js/line_under_header.js"></script>

      </div>
<footer class="footer">
  <div class="container">
    <div class="row">
      <div class="col-md-4 col-sm-4 text-center text-lg-left">
        &copy; Copyright Roman Hillje - All rights reserved
      </div>
      <section class="col-md-4 col-sm-4 text-center">
        <a target="_blank" href="https://www.twitter.com/fakechek1"><i class="fab fa-twitter" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a target="_blank" href="https://github.com/romanhaa"><i class="fab fa-github" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a target="_blank" href="https://gitlab.com/romanhaa"><i class="fab fa-gitlab" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a target="_blank" href="https://hub.docker.com/u/romanhaa"><i class="fab fa-docker" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a target="_blank" href="https://linkedin.com/in/roman.hillje"><i class="fab fa-linkedin" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
      </section>
      <div class="col-md-4 col-sm-4 text-center text-lg-right">
        Adapted from <a target="_blank" href="https://www.wowthemes.net">Mediumish Theme</a>
      </div>
    </div>
  </div>
</footer>
</div>

<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>
<script src="/js/mediumish.js"></script>
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script>
<script src="/js/toc.js"></script>

  </body>
</html>
