<!DOCTYPE html>
<html lang="en-us"><head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="generator" content="Hugo 0.57.2" />
  
  <title>Roman Hillje - Data Visualization &amp; Bioinformatics</title>
  
  
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.0.13/css/all.css" integrity="sha384-DNOHZ68U8hZfKXOrtjWvjxusGo9WQnrNx2sqG0tfsghAvtVlRW3tvkXWZh58N9jp" crossorigin="anonymous">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Righteous%7CMerriweather:300,300i,400,400i,700,700i">
  <link rel="stylesheet" href="/css/medium.css">
</head>
<body>
<nav class="navbar navbar-expand-lg navbar-light bg-white fixed-top mediumnavigation nav-down">
  <div class="container pr-0">

    
    <a class="navbar-brand" href="http://romanhaa.github.io/">
      
      <span style="font-family:Righteous;">Roman Hillje - Data Visualization &amp; Bioinformatics</span>
      
    </a>

    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarMediumish" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    
    <div class="collapse navbar-collapse" id="navbarMediumish">
      
      <ul class="navbar-nav ml-auto">
        
        <li class="nav-item ">
          <a class="nav-link" href="/">About Me</a>
        </li>
        
        <li class="nav-item ">
          <a class="nav-link" href="/blog">Blog</a>
        </li>
        
        <li class="nav-item ">
          <a class="nav-link" href="/plots">Plot Gallery</a>
        </li>
        
        <li class="nav-item ">
          <a class="nav-link" href="/projects">Projects</a>
        </li>
        
      </ul>
    </div>

  </div>
</nav>

<div class="site-content">
			<div class="container">
  <div class="main-content">

    
    <div class="container">
      <div class="row">

        
        <div class="col-md-12 flex-first flex-md-unordered">
          <div class="mainheading">

            
            <h1 class="posttitle">Scatter plot 4</h1>
            <div class="article-post">
              <p>Category: scatter</p>
            </div>
          </div>

          
          
            <img class="featured-image img-fluid" src="/images/plots/scatter/4.png" alt="thumbnail for this post" max-width="400">
          

          
          <div class="article-post">
            <pre><code class="language-r">library(tidyverse)
library(Seurat)

# load a single cell expression data set (generated in the lab I work at)
seurat &lt;- readRDS('seurat.rds')

# cells will be grouped by clusters that they have been assigned to
cluster_ids &lt;- levels(seurat@meta.data$seurat_clusters)

# select a set of genes for which we want to show expression
genes_to_show &lt;- seurat@misc$marker_genes$by_cluster %&gt;%
  group_by(cluster) %&gt;%
  arrange(-avg_logFC) %&gt;%
  slice(1) %&gt;%
  pull(gene)

# for every cluster-gene combination, calculate the average expression across
# all cells and then transform the data into a data frame
expression_levels_per_cluster &lt;- vapply(
    cluster_ids, FUN.VALUE = numeric(length(cluster_ids)), function(x) {
      cells_in_current_cluster &lt;- which(seurat@meta.data$seurat_cluster == x)
      Matrix::rowMeans(seurat@assays$SCT@data[genes_to_show,cells_in_current_cluster])
    }
  ) %&gt;%
  t() %&gt;%
  as.data.frame() %&gt;%
  mutate(cluster = rownames(.)) %&gt;%
  select(cluster, everything()) %&gt;%
  pivot_longer(
    cols = c(2:ncol(.)),
    names_to = 'gene'
  ) %&gt;%
  rename(expression = value) %&gt;%
  mutate(id_to_merge = paste0(cluster, '_', gene))

# for every cluster-gene combination, calculate the percentage of cells in the
# respective group that has at least 1 transcript (this means we consider it
# as expressing the gene) and then transform the data into a data frame
percentage_of_cells_expressing_gene &lt;- vapply(
    cluster_ids, FUN.VALUE = numeric(length(cluster_ids)), function(x) {
      cells_in_current_cluster &lt;- which(seurat@meta.data$seurat_cluster == x)
      Matrix::rowSums(seurat@assays$SCT@data[genes_to_show,cells_in_current_cluster] != 0)
    }
  ) %&gt;%
  t() %&gt;%
  as.data.frame() %&gt;%
  mutate(cluster = rownames(.)) %&gt;%
  select(cluster, everything()) %&gt;%
  pivot_longer(
    cols = c(2:ncol(.)),
    names_to = 'gene'
  ) %&gt;%
  rename(cell_count = value) %&gt;%
  left_join(
    .,
    seurat@meta.data %&gt;%
      group_by(seurat_clusters) %&gt;%
      tally() %&gt;%
      rename(cluster = seurat_clusters),
    by = 'cluster') %&gt;%
  mutate(
    id_to_merge = paste0(cluster, '_', gene),
    percent_cells = cell_count / n
  )

# merge the two data frames created before and plot the data
p &lt;- left_join(
    expression_levels_per_cluster,
    percentage_of_cells_expressing_gene %&gt;% select(id_to_merge, percent_cells),
    by = 'id_to_merge'
  ) %&gt;%
  mutate(cluster = factor(cluster, levels = rev(cluster_ids))) %&gt;%
  ggplot(aes(gene, cluster)) +
  geom_point(aes(color = expression, size = percent_cells)) +
  scale_color_distiller(
    palette = 'Reds',
    direction = 1,
    name = 'Log-normalised\nexpression',
    guide = guide_colorbar(frame.colour = &quot;black&quot;, ticks.colour = &quot;black&quot;)
  ) +
  scale_size(name = 'Percent\nof cells', labels = scales::percent) +
  labs(y = 'Cluster', color = 'Expression') +
  coord_fixed() +
  theme_bw() +
  theme(
    axis.title.x = element_blank(),
    axis.text.x = element_text(angle = 45, hjust = 1)
  )
ggsave('4.png', p, height = 7, width = 8)
</code></pre>

          </div>

        </div>
      </div>
    </div>
  </div>

			</div>
<footer class="footer">
  <div class="container">
    <div class="row">
      <div class="col-md-4 col-sm-4 text-center text-lg-left">
        &copy; Copyright Roman Hillje - All rights reserved
      </div>
      <section class="col-md-4 col-sm-4 text-center">
        <a target="_blank" href="https://github.com/romanhaa"><i class="fab fa-github" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a target="_blank" href="https://www.twitter.com/fakechek1"><i class="fab fa-twitter" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a target="_blank" href="https://linkedin.com/in/roman.hillje"><i class="fab fa-linkedin" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
      </section>
      <div class="col-md-4 col-sm-4 text-center text-lg-right">
        Adapted from <a target="_blank" href="https://www.wowthemes.net">Mediumish Theme</a>
      </div>
    </div>
  </div>
</footer>
</div>

<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>
<script src="/js/mediumish.js"></script>

	</body>
</html>