<!DOCTYPE html>
<html lang="en-us"><head>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
  <meta name="generator" content="Hugo 0.101.0" />
  
  <title>Roman Hillje - Programming &amp; Data Visualization</title>
  
  
  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
  <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.0.13/css/all.css" integrity="sha384-DNOHZ68U8hZfKXOrtjWvjxusGo9WQnrNx2sqG0tfsghAvtVlRW3tvkXWZh58N9jp" crossorigin="anonymous">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Righteous%7CMerriweather:300,300i,400,400i,700,700i">
  <link rel="stylesheet" href="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.css"/>
  <link rel="stylesheet" href="/css/medium.css">
</head>
<body><nav class="navbar navbar-expand-lg navbar-light bg-white mediumnavigation nav-down navbar-static-top">
  <div class="container pr-0">

    
    <a class="navbar-brand" href="http://romanhaa.github.io/">
      
      <span style="font-family:Righteous;">Roman Hillje - Programming &amp; Data Visualization</span>
      
    </a>

    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarMediumish" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    
    <div class="collapse navbar-collapse" id="navbarMediumish">
      
      <ul class="navbar-nav ml-auto">
        
        <li class="nav-item ">
          <a
            class="nav-link"
            href="/"
            
          >
            About Me
          </a>
        </li>
        
        <li class="nav-item ">
          <a
            class="nav-link"
            href="/projects"
            
          >
            Projects
          </a>
        </li>
        
        <li class="nav-item ">
          <a
            class="nav-link"
            href="/phd_thesis"
            
          >
            PhD Thesis
          </a>
        </li>
        
        <li class="nav-item ">
          <a
            class="nav-link"
            href="/plots"
            
          >
            Plot Gallery
          </a>
        </li>
        
        <li class="nav-item ">
          <a
            class="nav-link"
            href="/blog"
            
          >
            Blog
          </a>
        </li>
        
      </ul>
    </div>

  </div>
</nav>
<div class="site-content">
      <div class="container">
  <link rel="stylesheet" href="/css/numbered_headers.css">
  <div class="main-content">
    
    <div class="container">
      <div class="row">
        
        <div class="col-md-12 flex-first flex-md-unordered">
          <div class="mainheading">
            
            <h1 class="posttitle">Benchmark of scRNA-seq analysis tools</h1>
            <div class="article-post">
              Jun 23, 2020
            </div>
          </div>
          
          
          
          <body>
            <div class="article-post">
              <p>At the end of last year, I was curious to compare the results of a few single cell RNA-seq analysis tools, focusing on the generation of the transcript count matrix from the raw FASTQ files.
The contenders are: <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger" target="_blank" rel="noreferrer noopener">10x Genomics Cell Ranger</a>, <a href="https://github.com/alexdobin/STAR" target="_blank" rel="noreferrer noopener">STARsolo</a>, <a href="https://salmon.readthedocs.io/en/latest/alevin.html" target="_blank" rel="noreferrer noopener">Alevin (Salmon)</a>, and <a href="https://pachterlab.github.io/kallisto/about" target="_blank" rel="noreferrer noopener">kallisto</a>.
Even though this was done a while ago, and updates of the tools possibly influence the results, I thought it might be worth sharing anyway.</p>
<h2 id="preparation">Preparation</h2>
<h3 id="download-raw-data">Download raw data</h3>
<p>I only used a single data set, which consists of roughly 11,000 cells from a human PBMC sample, was generated using the 10x Genomics Single Cell Gene Expression kit (v3 chemistry), and can be downloaded from the <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/pbmc_10k_v3" target="_blank" rel="noreferrer noopener">10x Genomics website</a>.</p>
<p>The data set contains 639 million reads, resulting in a sequencing depth of approximately 54,000 reads per cell.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>mkdir -p &lt;project_home&gt;/raw_data/pbmc_10k_v3
</span></span><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/raw_data/pbmc_10k_v3/
</span></span><span style="display:flex;"><span>wget http://s3-us-west-2.amazonaws.com/10x.files/samples/cell-exp/3.0.0/pbmc_10k_v3/pbmc_10k_v3_fastqs.tar
</span></span><span style="display:flex;"><span>tar -xf pbmc_10k_v3_fastqs.tar
</span></span><span style="display:flex;"><span>mv pbmc_10k_v3_fastqs FASTQ
</span></span></code></pre></div><h3 id="collect-reference">Collect reference</h3>
<p>All reference files (genome as <code>FASTA</code>, transcriptome as <code>FASTA</code>, and annotation as <code>GTF</code>) were taken from <a href="https://www.gencodegenes.org/human/" target="_blank" rel="noreferrer noopener">GENCODE (v32)</a>.</p>
<h4 id="genome">Genome</h4>
<p>We use the <code>hg38</code> release of the human genome.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>mkdir -p &lt;project_home&gt;/reference_files/hg38
</span></span><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/reference_files/hg38
</span></span><span style="display:flex;"><span>wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/GRCh38.primary_assembly.genome.fa.gz
</span></span><span style="display:flex;"><span>gunzip GRCh38.primary_assembly.genome.fa.gz
</span></span></code></pre></div><h4 id="transcriptome">Transcriptome</h4>
<p>We use all transcripts.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/reference_files/hg38
</span></span><span style="display:flex;"><span>wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.transcripts.fa.gz
</span></span><span style="display:flex;"><span>gunzip gencode.v32.transcripts.fa.gz
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>grep -c <span style="color:#d14">&#39;^&gt;ENST&#39;</span> gencode.v32.transcripts.fa
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># 227,462</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># list of transcripts</span>
</span></span><span style="display:flex;"><span>cut -f1 gencode.v32.annotation_tgMap.tsv &gt; gencode.v32.annotation_transcripts.tsv
</span></span></code></pre></div><h4 id="annotation">Annotation</h4>
<p>We use the comprehensive annotation.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/reference_files/hg38
</span></span><span style="display:flex;"><span>wget ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_32/gencode.v32.annotation.gtf.gz
</span></span><span style="display:flex;"><span>gunzip gencode.v32.annotation.gtf.gz
</span></span></code></pre></div><h4 id="transcript-to-gene-map-file-for-salmon">Transcript to gene map file for Salmon</h4>
<p>Alevin annd kallisto require a <code>TSV</code> file that connect every annotated transcript to its corresponding gene ID.
We can generate this file from the previously downloaded <code>GTF</code> file containing the comprehensive GENCODE annotation.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#900;font-weight:bold">setwd</span>(<span style="color:#d14">&#39;&lt;project_home&gt;/&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">library</span>(tidyverse)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>gtf <span style="color:#000;font-weight:bold">&lt;-</span> rtracklayer<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">import</span>(<span style="color:#d14">&#39;reference_files/hg38/gencode.v32.annotation.gtf&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>tgMap <span style="color:#000;font-weight:bold">&lt;-</span> gtf <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as_tibble</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">select</span>(transcript_id,gene_id) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">filter</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#000;font-weight:bold">!</span><span style="color:#900;font-weight:bold">is.na</span>(transcript_id),
</span></span><span style="display:flex;"><span>    <span style="color:#000;font-weight:bold">!</span><span style="color:#900;font-weight:bold">is.na</span>(gene_id)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">distinct</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">write_tsv</span>(tgMap, <span style="color:#d14">&#39;reference_files/hg38/gencode.v32.annotation_tgMap.tsv&#39;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>)
</span></span></code></pre></div><h4 id="list-of-cellular-barcodes-of-10x-v3-kit">List of cellular barcodes of 10x v3 kit</h4>
<p>For some steps of the analysis, we will need a list of all valid cellular barcodes in the 10x Genomics Single Cell Gene Expression kit (v3 chemistry).
It&rsquo;s well hidden but can be found in the Cell Ranger pipeline installation (<code>cellranger-x.y.z/cellranger-cs/x.y.z/lib/python/cellranger/barcodes/3M-february-2018.txt.gz</code>).
I previously extracted this file and unpacked it, so here I just take it from there.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/
</span></span><span style="display:flex;"><span>mkdir &lt;project_home&gt;/software
</span></span><span style="display:flex;"><span>cp &lt;some_other_folder&gt;/3M-february-2018.txt &lt;project_home&gt;/software
</span></span></code></pre></div><h3 id="get-tools">Get tools</h3>
<p>I have created separate Docker containers with Cell Ranger, STARsolo, Alevin (Salmon), and kallisto installed which I import into Singularity.</p>
<h4 id="cell-ranger">Cell Ranger</h4>
<p>We use Cell Ranger 3.1.0.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>singularity build &lt;project_home&gt;/software/cell_ranger_3.1.0.simg docker://romanhaa/cell_ranger:3.1.0
</span></span></code></pre></div><h4 id="starsolo">STARsolo</h4>
<p>We use STAR 2.7.3a.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>singularity build &lt;project_home&gt;/software/star_2.7.3a.simg docker://romanhaa/star:2.7.3a
</span></span></code></pre></div><h4 id="alevin-salmon">Alevin (Salmon)</h4>
<p>We use Alevin 0.14.1.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>singularity build &lt;project_home&gt;/software/salmon_0.14.1.simg docker://combinelab/salmon:0.14.1
</span></span></code></pre></div><h4 id="kallisto">kallisto</h4>
<p>We use kallisto 0.46.0.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>singularity build &lt;project_home&gt;/software/kallist_0.46.0.simg docker://romanhaa/kallisto:0.46.0
</span></span></code></pre></div><h3 id="create-indices">Create indices</h3>
<h4 id="cell-ranger-1">Cell Ranger</h4>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/reference_files/hg38
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#0086b3">exec</span> --bind /hpcnfs &lt;project_home&gt;/software/cell_ranger_3.1.0.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>cellranger mkref <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--genome<span style="color:#000;font-weight:bold">=</span>cell_ranger_3.1.0 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--fasta<span style="color:#000;font-weight:bold">=</span>GRCh38.primary_assembly.genome.fa <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--genes<span style="color:#000;font-weight:bold">=</span>gencode.v32.annotation.gtf <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--nthreads<span style="color:#000;font-weight:bold">=</span><span style="color:#099">10</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--memgb<span style="color:#000;font-weight:bold">=</span><span style="color:#099">8</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># /cellranger-3.1.0/cellranger-cs/3.1.0/bin</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># cellranger mkref (3.1.0)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Copyright (c) 2019 10x Genomics, Inc.  All rights reserved.</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># -------------------------------------------------------------------------------</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Creating new reference folder at &lt;project_home&gt;/cell_ranger_3.1.0</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Writing genome FASTA file into reference folder...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Computing hash of genome FASTA file...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Indexing genome FASTA file...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Writing genes GTF file into reference folder...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Computing hash of genes GTF file...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Writing genes index file into reference folder (may take over 10 minutes for a 3Gb genome)...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Writing genome metadata JSON file into reference folder...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Generating STAR genome index (may take over 8 core hours for a 3Gb genome)...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:33:36 ..... Started STAR run</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:33:37 ... Starting to generate Genome files</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:35:25 ... starting to sort  Suffix Array. This may take a long time...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:35:29 ... sorting Suffix Array chunks and saving them to disk...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:36:11 ... loading chunks from disk, packing SA...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:36:43 ... Finished generating suffix array</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:36:43 ... Generating Suffix Array index</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:37:23 ... Completed Suffix Array index</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:37:23 ..... Processing annotations GTF</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:37:53 ..... Inserting junctions into the genome indices</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:40:54 ... writing Genome to disk ...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:41:38 ... writing Suffix Array to disk ...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:41:54 ... writing SAindex to disk</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:42:14 ..... Finished successfully</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># ...done.</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># &gt;&gt;&gt; Reference successfully created! &lt;&lt;&lt;</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># You can now specify this reference on the command line:</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># cellranger --transcriptome=&lt;project_home&gt;/reference_files/hg38/cell_ranger_3.1.0 ...</span>
</span></span></code></pre></div><h4 id="starsolo-1">STARsolo</h4>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/reference_files/hg38
</span></span><span style="display:flex;"><span>mkdir STAR_2.7.3a
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#0086b3">exec</span> --bind /hpcnfs &lt;project_home&gt;/software/star_2.7.3a.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>STAR <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--runThreadN <span style="color:#099">12</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--runMode genomeGenerate <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--genomeDir STAR_2.7.3a <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--genomeFastaFiles GRCh38.primary_assembly.genome.fa <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--sjdbGTFfile gencode.v32.annotation.gtf <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--sjdbGTFtagExonParentGene gene_id <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--sjdbOverhang <span style="color:#099">90</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:16:56 ..... started STAR run</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:16:56 ... starting to generate Genome files</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:18:01 ... starting to sort Suffix Array. This may take a long time...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:18:17 ... sorting Suffix Array chunks and saving them to disk...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:46:47 ... loading chunks from disk, packing SA...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:56:06 ... finished generating suffix array</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 13:56:06 ... generating Suffix Array index</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 14:00:04 ... completed Suffix Array index</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 14:00:04 ..... processing annotations GTF</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 14:00:32 ..... inserting junctions into the genome indices</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 14:03:41 ... writing Genome to disk ...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 14:04:10 ... writing Suffix Array to disk ...</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 14:07:46 ... writing SAindex to disk</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Oct 26 14:08:02 ..... finished successfully</span>
</span></span></code></pre></div><h4 id="alevin-salmon-1">Alevin (Salmon)</h4>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;
</span></span><span style="display:flex;"><span>mkdir reference_files/hg38/salmon_index_k_31
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#0086b3">exec</span> --bind /hpcnfs &lt;project_home&gt;/software/salmon_0.14.1.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>salmon index <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--transcripts reference_files/hg38/gencode.v32.transcripts.fa <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--gencode <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--index reference_files/hg38/salmon_index_k_31 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--kmerLen <span style="color:#099">31</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Version Info: Could not resolve upgrade information in the alotted time.</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Check for upgrades manually at https://combine-lab.github.io/salmon</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># index [&#34;reference_files/hg38/salmon_index_k_31&#34;] did not previously exist  . . . creating it</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:46.039] [jLog] [info] building index</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:46.043] [jointLog] [info] [Step 1 of 4] : counting k-mers</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:48.050] [jointLog] [warning] Entry with header [ENST00000473810.1|ENSG00000239255.1|OTTHUMG00000157482.1|OTTHUMT00000348942.1|AC007620.1-201|AC007620.1|25|processed_pseudogene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:48.178] [jointLog] [warning] Entry with header [ENST00000603775.1|ENSG00000271544.1|OTTHUMG00000184300.1|OTTHUMT00000468575.1|AC006499.8-201|AC006499.8|23|processed_pseudogene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:49.680] [jointLog] [warning] Entry with header [ENST00000632684.1|ENSG00000282431.1|OTTHUMG00000190602.2|OTTHUMT00000485301.2|TRBD1-202|TRBD1|12|TR_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:51.483] [jointLog] [warning] Entry with header [ENST00000626826.1|ENSG00000281344.1|OTTHUMG00000189570.1|OTTHUMT00000479989.1|HELLPAR-201|HELLPAR|205012|lncRNA|] was longer than 200000 nucleotides.  Are you certain that we are indexing a transcriptome and not a genome?</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:51.558] [jointLog] [warning] Entry with header [ENST00000543745.1|ENSG00000255972.1|OTTHUMG00000168883.1|OTTHUMT00000401485.1|AC026333.1-201|AC026333.1|28|processed_pseudogene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:51.777] [jointLog] [warning] Entry with header [ENST00000415118.1|ENSG00000223997.1|OTTHUMG00000170844.2|OTTHUMT00000410670.2|TRDD1-201|TRDD1|8|TR_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:51.777] [jointLog] [warning] Entry with header [ENST00000434970.2|ENSG00000237235.2|OTTHUMG00000170845.2|OTTHUMT00000410671.2|TRDD2-201|TRDD2|9|TR_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:51.777] [jointLog] [warning] Entry with header [ENST00000448914.1|ENSG00000228985.1|OTTHUMG00000170846.2|OTTHUMT00000410672.2|TRDD3-201|TRDD3|13|TR_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.027] [jointLog] [warning] Entry with header [ENST00000439842.1|ENSG00000236597.1|OTTHUMG00000152435.2|OTTHUMT00000326213.2|IGHD7-27-201|IGHD7-27|11|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.027] [jointLog] [warning] Entry with header [ENST00000390567.1|ENSG00000211907.1|OTTHUMG00000152429.2|OTTHUMT00000326207.2|IGHD1-26-201|IGHD1-26|20|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.027] [jointLog] [warning] Entry with header [ENST00000452198.1|ENSG00000225825.1|OTTHUMG00000152436.2|OTTHUMT00000326214.2|IGHD6-25-201|IGHD6-25|18|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.027] [jointLog] [warning] Entry with header [ENST00000390569.1|ENSG00000211909.1|OTTHUMG00000152427.2|OTTHUMT00000326205.2|IGHD5-24-201|IGHD5-24|20|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.027] [jointLog] [warning] Entry with header [ENST00000437320.1|ENSG00000227196.1|OTTHUMG00000152438.2|OTTHUMT00000326216.2|IGHD4-23-201|IGHD4-23|19|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.027] [jointLog] [warning] Entry with header [ENST00000390572.1|ENSG00000211912.1|OTTHUMG00000152428.2|OTTHUMT00000326206.2|IGHD2-21-201|IGHD2-21|28|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000450276.1|ENSG00000237020.1|OTTHUMG00000152432.2|OTTHUMT00000326210.2|IGHD1-20-201|IGHD1-20|17|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000390574.1|ENSG00000211914.1|OTTHUMG00000152431.2|OTTHUMT00000326209.2|IGHD6-19-201|IGHD6-19|21|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000390575.1|ENSG00000211915.1|OTTHUMG00000152433.2|OTTHUMT00000326211.2|IGHD5-18-201|IGHD5-18|20|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000431870.1|ENSG00000227800.1|OTTHUMG00000152437.2|OTTHUMT00000326215.2|IGHD4-17-201|IGHD4-17|16|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000451044.1|ENSG00000227108.1|OTTHUMG00000152369.2|OTTHUMT00000326003.2|IGHD1-14-201|IGHD1-14|17|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000390580.1|ENSG00000211920.1|OTTHUMG00000152370.2|OTTHUMT00000326004.2|IGHD6-13-201|IGHD6-13|21|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000390581.1|ENSG00000211921.1|OTTHUMG00000152367.2|OTTHUMT00000326001.2|IGHD5-12-201|IGHD5-12|23|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000431440.2|ENSG00000232543.2|OTTHUMG00000152368.2|OTTHUMT00000326002.2|IGHD4-11-201|IGHD4-11|16|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000430425.1|ENSG00000237197.1|OTTHUMG00000152357.2|OTTHUMT00000325963.2|IGHD1-7-201|IGHD1-7|17|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000454691.1|ENSG00000228131.1|OTTHUMG00000152353.2|OTTHUMT00000325959.2|IGHD6-6-201|IGHD6-6|18|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000390588.1|ENSG00000211928.1|OTTHUMG00000152360.2|OTTHUMT00000325966.2|IGHD5-5-201|IGHD5-5|20|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000414852.1|ENSG00000233655.1|OTTHUMG00000152355.2|OTTHUMT00000325961.2|IGHD4-4-201|IGHD4-4|16|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.028] [jointLog] [warning] Entry with header [ENST00000454908.1|ENSG00000236170.1|OTTHUMG00000152359.2|OTTHUMT00000325965.2|IGHD1-1-201|IGHD1-1|17|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.029] [jointLog] [warning] Entry with header [ENST00000518246.1|ENSG00000254045.1|OTTHUMG00000152060.1|OTTHUMT00000325154.1|IGHVIII-22-2-201|IGHVIII-22-2|28|IG_V_pseudogene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.030] [jointLog] [warning] Entry with header [ENST00000604642.1|ENSG00000270961.1|OTTHUMG00000184622.2|OTTHUMT00000468982.2|IGHD5OR15-5A-201|IGHD5OR15-5A|23|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.030] [jointLog] [warning] Entry with header [ENST00000603326.1|ENSG00000271317.1|OTTHUMG00000184621.3|OTTHUMT00000468981.3|IGHD4OR15-4A-201|IGHD4OR15-4A|19|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.030] [jointLog] [warning] Entry with header [ENST00000605284.1|ENSG00000271336.1|OTTHUMG00000184580.2|OTTHUMT00000468908.2|IGHD1OR15-1A-201|IGHD1OR15-1A|17|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.033] [jointLog] [warning] Entry with header [ENST00000604446.1|ENSG00000270824.1|OTTHUMG00000184624.2|OTTHUMT00000468984.2|IGHD5OR15-5B-201|IGHD5OR15-5B|23|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.042] [jointLog] [warning] Entry with header [ENST00000603693.1|ENSG00000270451.1|OTTHUMG00000184611.3|OTTHUMT00000468945.3|IGHD4OR15-4B-201|IGHD4OR15-4B|19|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.042] [jointLog] [warning] Entry with header [ENST00000604838.1|ENSG00000270185.1|OTTHUMG00000184585.2|OTTHUMT00000468915.2|IGHD1OR15-1B-201|IGHD1OR15-1B|17|IG_D_gene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:52.837] [jointLog] [warning] Entry with header [ENST00000579054.1|ENSG00000266416.1|OTTHUMG00000179204.1|OTTHUMT00000445280.1|AC130289.2-201|AC130289.2|28|processed_pseudogene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:53.480] [jointLog] [warning] Entry with header [ENST00000634174.1|ENSG00000282732.1|OTTHUMG00000191398.1|OTTHUMT00000487783.1|AC073539.7-201|AC073539.7|28|unprocessed_pseudogene|], had length less than the k-mer length of 31 (perhaps after poly-A clipping)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Elapsed time: 8.59662s</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:54.640] [jointLog] [warning] Removed 854 transcripts that were sequence duplicates of indexed transcripts.</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:54.640] [jointLog] [warning] If you wish to retain duplicate transcripts, please use the `--keepDuplicates` flag</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:54.657] [jointLog] [info] Replaced 4 non-ATCG nucleotides</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:54.657] [jointLog] [info] Clipped poly-A tails from 1814 transcripts</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:54.671] [jointLog] [info] Building rank-select dictionary and saving to disk</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:54.695] [jointLog] [info] done</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Elapsed time: 0.0236391s</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:55.111] [jointLog] [info] Writing sequence data to file . . .</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:55.329] [jointLog] [info] done</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Elapsed time: 0.218913s</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:58.504] [jointLog] [info] Building 32-bit suffix array (length of generalized text is 356944333)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:48:58.943] [jointLog] [info] Building suffix array . . .</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># success</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># saving to disk . . . done</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Elapsed time: 0.843981s</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># done</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Elapsed time: 44.8151s</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># processed 356000000 positions[2019-10-24 09:52:10.475] [jointLog] [info] khash had 143394783 keys</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:52:10.479] [jointLog] [info] saving hash to disk . . .</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:52:19.074] [jointLog] [info] done</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># Elapsed time: 8.59516s</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [2019-10-24 09:52:40.740] [jLog] [info] done building index</span>
</span></span></code></pre></div><h4 id="kallisto-1">kallisto</h4>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>singularity <span style="color:#0086b3">exec</span> --bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>software/kallisto_0.46.0.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>kallisto index <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--index reference_files/hg38/kallisto_index_k_31 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--kmer-size <span style="color:#099">31</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>reference_files/hg38/gencode.v32.transcripts.fa
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [build] loading fasta file reference_files/hg38/gencode.v32.transcripts.fa</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [build] k-mer length: 31</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [build] warning: clipped off poly-A tail (longer than 10)</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#         from 1814 target sequences</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [build] warning: replaced 4 non-ACGUT characters in the input sequence</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#         with pseudorandom nucleotides</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [build] counting k-mers ... done.</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [build] building target de Bruijn graph ...  done</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [build] creating equivalence classes ...  done</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># [build] target de Bruijn graph has 1472677 contigs and contains 140115246 k-mers</span>
</span></span></code></pre></div><p><strong>NOTE:</strong> Similar to Alevin, this step took about 5 minutes to finish.</p>
<h2 id="quantify-gene-expression">Quantify gene expression</h2>
<p>Every tool has its own way of operation, which means that Alevin needs to be run multiple times in order to generate output that can be compared with the other tools.</p>
<p>Each job will be run on a single node of our HPC, with 20 threads/cores and 128GB of memory available.
No other jobs compete for the same memory.</p>
<h3 id="cell-ranger-2">Cell Ranger</h3>
<p>Cell Ranger outputs both the raw and a filtered matrix of transcript counts.
The filtered matrix contains only those cellular barcodes that have been identified as belonging to actual cells, whereas the others are background and therefore removed.</p>
<p>Cell Ranger might not be the best tool for the analysis out there, but since it is provided by 10x Genomics, the manufacturer of the library preparation kit, we will use the list of cells identified by it, also known as whitelist, to compare the tools.</p>
<p>Here is the script that was submitted to the job scheduler.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -N cell_ranger</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -l select=1:ncpus=28:mem=120gb</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -o &lt;..&gt;/sh_files/cell_ranger.stdout</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -e &lt;..&gt;/sh_files/cell_ranger.stderr</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -m ae</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -M roman.hillje@ieo.it</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>mkdir &lt;project_home&gt;/results/pbmc_10k_v3/cell_ranger
</span></span><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/results/pbmc_10k_v3/cell_ranger
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span><span style="color:#0086b3">exec</span> --bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>&lt;project_home&gt;/software/cell_ranger_3.1.0.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>cellranger count <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--id<span style="color:#000;font-weight:bold">=</span>pbmc_10k_v3 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--transcriptome<span style="color:#000;font-weight:bold">=</span>&lt;project_home&gt;/reference_files/hg38/cell_ranger_3.1.0 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--fastqs<span style="color:#000;font-weight:bold">=</span>&lt;project_home&gt;/raw_data/pbmc_10k_v3/FASTQ <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--sample<span style="color:#000;font-weight:bold">=</span>pbmc_10k_v3 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--expect-cells<span style="color:#000;font-weight:bold">=</span><span style="color:#099">10000</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--jobmode<span style="color:#000;font-weight:bold">=</span><span style="color:#0086b3">local</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--localcores<span style="color:#000;font-weight:bold">=</span><span style="color:#099">20</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--localmem<span style="color:#000;font-weight:bold">=</span><span style="color:#099">100</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--mempercore<span style="color:#000;font-weight:bold">=</span><span style="color:#099">5</span>
</span></span></code></pre></div><p>Below is a summary of the run time and used resources.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>resources_used.cpupercent<span style="color:#000;font-weight:bold">=</span><span style="color:#099">1077</span>
</span></span><span style="display:flex;"><span>resources_used.cput<span style="color:#000;font-weight:bold">=</span>222:30:36
</span></span><span style="display:flex;"><span>resources_used.mem<span style="color:#000;font-weight:bold">=</span>125829116kb
</span></span><span style="display:flex;"><span>resources_used.ncpus<span style="color:#000;font-weight:bold">=</span><span style="color:#099">28</span>
</span></span><span style="display:flex;"><span>resources_used.vmem<span style="color:#000;font-weight:bold">=</span>125829120kb
</span></span><span style="display:flex;"><span>resources_used.walltime<span style="color:#000;font-weight:bold">=</span>20:39:07
</span></span></code></pre></div><h3 id="starsolo-2">STARsolo</h3>
<p>Also STAR automatically generates a raw and filtered transcript count matrix.
The algorithm for cellular barcode filtering is one that was used in previous Cell Ranger versions.</p>
<p>Here is the script that was submitted to the job scheduler.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -N star</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -l select=1:ncpus=28:mem=120gb</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -o &lt;project_home&gt;/sh_files/star.stdout</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -e &lt;project_home&gt;/sh_files/star.stderr</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -m ae</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -M roman.hillje@ieo.it</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>mkdir &lt;project_home&gt;/results/pbmc_10k_v3/STAR
</span></span><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;/results/pbmc_10k_v3/STAR
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span><span style="color:#0086b3">exec</span> --bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>&lt;project_home&gt;/software/star_2.7.3a.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>STAR <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--runThreadN <span style="color:#099">20</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--genomeDir &lt;project_home&gt;/reference_files/hg38/STAR_2.7.3a <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--readFilesIn <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>&lt;project_home&gt;/raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L001_R2_001.fastq.gz,&lt;project_home&gt;/raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L002_R2_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>&lt;project_home&gt;/raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L001_R1_001.fastq.gz,&lt;project_home&gt;/raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L002_R1_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--readFilesCommand zcat <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--outFileNamePrefix ./ <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--outTmpKeep None <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--outStd Log <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--outSAMtype None <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloType CB_UMI_Simple <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloCBwhitelist &lt;project_home&gt;/software/3M-february-2018.txt <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloCBstart <span style="color:#099">1</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloCBlen <span style="color:#099">16</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloUMIstart <span style="color:#099">17</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloUMIlen <span style="color:#099">12</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloBarcodeReadLength <span style="color:#099">1</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloCBmatchWLtype 1MM_multi_pseudocounts <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloStrand Forward <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloFeatures Gene <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloUMIdedup 1MM_All <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloUMIfiltering MultiGeneUMI <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloOutFileNames STARsolo/ features.tsv barcodes.tsv matrix.mtx <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--soloCellFilter CellRanger2.2 <span style="color:#099">3000</span> 0.99 <span style="color:#099">10</span>
</span></span></code></pre></div><p>Below is a summary of the run time and used resources.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>resources_used.cpupercent<span style="color:#000;font-weight:bold">=</span><span style="color:#099">1584</span>
</span></span><span style="display:flex;"><span>resources_used.cput<span style="color:#000;font-weight:bold">=</span>09:10:46
</span></span><span style="display:flex;"><span>resources_used.mem<span style="color:#000;font-weight:bold">=</span>45942620kb
</span></span><span style="display:flex;"><span>resources_used.ncpus<span style="color:#000;font-weight:bold">=</span><span style="color:#099">28</span>
</span></span><span style="display:flex;"><span>resources_used.vmem<span style="color:#000;font-weight:bold">=</span>46475712kb
</span></span><span style="display:flex;"><span>resources_used.walltime<span style="color:#000;font-weight:bold">=</span>00:34:45
</span></span></code></pre></div><h3 id="alevin-salmon-2">Alevin (Salmon)</h3>
<p>Unfortunately, at least at the time I performed this analysis, Alevin requires providing a list of cellular barcodes in order to run or otherwise it will apply its own filtering algorithm.
That means, to compare the tools, we have to run it twice:</p>
<ol>
<li>With the entire list of valid barcodes to generate the raw transcript count matrix.</li>
<li>Not providing any list and using the Alevin filtering algorithm.</li>
</ol>
<h4 id="with-all-cellular-barcodes-as-whitelist-file-raw">With all cellular barcodes as whitelist file (raw)</h4>
<p>Here is the script that was submitted to the job scheduler.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -N alevin_1</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -l select=1:ncpus=28:mem=120gb</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -o &lt;project_home&gt;/sh_files/alevin_1.stdout</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -e &lt;project_home&gt;/sh_files/alevin_1.stderr</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -m ae</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -M roman.hillje@ieo.it</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#0086b3">exec</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>software/salmon_0.14.1.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>salmon alevin <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--libType ISR <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--mates1 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L001_R1_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L002_R1_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--mates2 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L001_R2_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L002_R2_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--chromiumV3 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--index reference_files/hg38/salmon_index_k_31 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--threads <span style="color:#099">20</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--dumpMtx <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--output results/pbmc_10k_v3/alevin/raw <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--whitelist software/3M-february-2018.txt <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--tgMap reference_files/hg38/gencode.v32.annotation_tgMap.tsv
</span></span></code></pre></div><p>Below is a summary of the run time and used resources.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>resources_used.cpupercent<span style="color:#000;font-weight:bold">=</span><span style="color:#099">1898</span>
</span></span><span style="display:flex;"><span>resources_used.cput<span style="color:#000;font-weight:bold">=</span>53:05:08
</span></span><span style="display:flex;"><span>resources_used.mem<span style="color:#000;font-weight:bold">=</span>21223076kb
</span></span><span style="display:flex;"><span>resources_used.ncpus<span style="color:#000;font-weight:bold">=</span><span style="color:#099">28</span>
</span></span><span style="display:flex;"><span>resources_used.vmem<span style="color:#000;font-weight:bold">=</span>34565292kb
</span></span><span style="display:flex;"><span>resources_used.walltime<span style="color:#000;font-weight:bold">=</span>03:09:25
</span></span></code></pre></div><h4 id="without-specified-whitelist-file-filtered">Without specified whitelist file (filtered)</h4>
<p>Here is the script that was submitted to the job scheduler.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -N alevin_2</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -l select=1:ncpus=28:mem=120gb</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -o &lt;project_home&gt;/sh_files/alevin_2.stdout</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -e &lt;project_home&gt;/sh_files/alevin_2.stderr</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -m ae</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -M roman.hillje@ieo.it</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#0086b3">exec</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>software/salmon_0.14.1.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>salmon alevin <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--libType ISR <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--mates1 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L001_R1_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L002_R1_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--mates2 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L001_R2_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L002_R2_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--chromiumV3 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--index reference_files/hg38/salmon_index_k_31 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--threads <span style="color:#099">20</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--dumpMtx <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--output results/pbmc_10k_v3/alevin/filtered <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--tgMap reference_files/hg38/gencode.v32.annotation_tgMap.tsv
</span></span></code></pre></div><p>Below is a summary of the run time and used resources.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>resources_used.cpupercent<span style="color:#000;font-weight:bold">=</span><span style="color:#099">1829</span>
</span></span><span style="display:flex;"><span>resources_used.cput<span style="color:#000;font-weight:bold">=</span>13:06:48
</span></span><span style="display:flex;"><span>resources_used.mem<span style="color:#000;font-weight:bold">=</span>18361588kb
</span></span><span style="display:flex;"><span>resources_used.ncpus<span style="color:#000;font-weight:bold">=</span><span style="color:#099">28</span>
</span></span><span style="display:flex;"><span>resources_used.vmem<span style="color:#000;font-weight:bold">=</span>33971372kb
</span></span><span style="display:flex;"><span>resources_used.walltime<span style="color:#000;font-weight:bold">=</span>00:59:45
</span></span></code></pre></div><h3 id="kallisto-2">kallisto</h3>
<p>kallisto does not have any integrated algorithm for cellular barcode filtering and expects you to provide a list of barcodes you want to analyze.
We simply run kallisto once to quantify transcript counts in all valid cellular barcodes and then extract the cells identified by Cell Ranger later.</p>
<p>Here is the script that was submitted to the job scheduler.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -N kallisto</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -l select=1:ncpus=28:mem=120gb</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -o &lt;project_home&gt;/sh_files/kallisto.stdout</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -e &lt;project_home&gt;/sh_files/kallisto.stderr</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -m ae</span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">#PBS -M roman.hillje@ieo.it</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#0086b3">cd</span> &lt;project_home&gt;
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span><span style="color:#0086b3">exec</span> --bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>software/kallisto_0.46.0.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>kallisto bus <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--index reference_files/hg38/kallisto_index_k_31 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--output-dir results/pbmc_10k_v3/kallisto <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--technology 10xv3 <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--threads <span style="color:#099">20</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L001_R1_001.fastq.gz raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L001_R2_001.fastq.gz <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L002_R1_001.fastq.gz raw_data/pbmc_10k_v3/FASTQ/pbmc_10k_v3_S1_L002_R2_001.fastq.gz
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#0086b3">cd</span> results/pbmc_10k_v3/kallisto
</span></span><span style="display:flex;"><span>mkdir genecount/ tmp/
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>singularity <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span><span style="color:#0086b3">exec</span> --bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>&lt;project_home&gt;/software/kallisto_0.46.0.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>bustools correct <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--whitelist &lt;project_home&gt;/software/3M-february-2018.txt <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--pipe <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>output.bus | <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>singularity <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span><span style="color:#0086b3">exec</span> --bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>&lt;project_home&gt;/software/kallisto_0.46.0.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>bustools sort <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--temp tmp/ <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--threads <span style="color:#099">20</span> <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--pipe <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>- | <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>singularity <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span><span style="color:#0086b3">exec</span> --bind /hpcnfs <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>&lt;project_home&gt;/software/kallisto_0.46.0.simg <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>bustools count <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--output genecount/genes <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--genemap &lt;project_home&gt;/reference_files/hg38/gencode.v32.annotation_tgMap.tsv <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--ecmap matrix.ec <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--txnames &lt;project_home&gt;/reference_files/hg38/gencode.v32.annotation_transcripts.tsv <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>--genecounts <span style="color:#d14">\
</span></span></span><span style="display:flex;"><span><span style="color:#d14"></span>-
</span></span></code></pre></div><p>Below is a summary of the run time and used resources.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-sh" data-lang="sh"><span style="display:flex;"><span>resources_used.cpupercent<span style="color:#000;font-weight:bold">=</span><span style="color:#099">404</span>
</span></span><span style="display:flex;"><span>resources_used.cput<span style="color:#000;font-weight:bold">=</span>01:54:21
</span></span><span style="display:flex;"><span>resources_used.mem<span style="color:#000;font-weight:bold">=</span>13737072kb
</span></span><span style="display:flex;"><span>resources_used.ncpus<span style="color:#000;font-weight:bold">=</span><span style="color:#099">28</span>
</span></span><span style="display:flex;"><span>resources_used.vmem<span style="color:#000;font-weight:bold">=</span>14877188kb
</span></span><span style="display:flex;"><span>resources_used.walltime<span style="color:#000;font-weight:bold">=</span>00:35:09
</span></span></code></pre></div><h2 id="compare-results">Compare results</h2>
<p>After producing all the transcript count matrices, it is time for some comparisons.</p>
<h3 id="load-data">Load data</h3>
<p>First, we prepare an object that will hold all the transcript counts.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#900;font-weight:bold">library</span>(tidyverse)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">list</span>()
</span></span></code></pre></div><p>Now, we can load and add the matrices one-by-one.
Importantly, for each raw transcript count matrix, we also (1) remove all genes that are not expressed in any cell and (2) cellular barcodes without any transcript in any gene (0 transcripts in total).
Instead, in the filtered matrices, we only remove the genes that are not expressed in any cell but leave the cellular barcodes as they are.</p>
<h4 id="cell-ranger-3">Cell Ranger</h4>
<h5 id="all-cellular-barcodes">All cellular barcodes</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">readMM</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/cell_ranger/pbmc_10k_v3/outs/raw_feature_bc_matrix/matrix.mtx.gz&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">colnames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/cell_ranger/pbmc_10k_v3/outs/raw_feature_bc_matrix/barcodes.tsv.gz&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">gsub</span>(pattern <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;-1&#39;</span>, replacement <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">rownames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/cell_ranger/pbmc_10k_v3/outs/raw_feature_bc_matrix/features.tsv.gz&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span>  
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove empty cells</span>
</span></span><span style="display:flex;"><span>empty_cells <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(temp)
</span></span><span style="display:flex;"><span>empty_cells <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(empty_cells <span style="color:#000;font-weight:bold">==</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[,<span style="color:#000;font-weight:bold">-</span>empty_cells]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_all_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h5 id="filtered-cellular-barcodes">Filtered cellular barcodes</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">readMM</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/cell_ranger/pbmc_10k_v3/outs/filtered_feature_bc_matrix/matrix.mtx.gz&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">colnames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/cell_ranger/pbmc_10k_v3/outs/filtered_feature_bc_matrix/barcodes.tsv.gz&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">gsub</span>(pattern <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;-1&#39;</span>, replacement <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">rownames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/cell_ranger/pbmc_10k_v3/outs/filtered_feature_bc_matrix/features.tsv.gz&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">make.unique</span>(sep <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#34;_&#34;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h4 id="starsolo-3">STARsolo</h4>
<h5 id="all-cellular-barcodes-1">All cellular barcodes</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">readMM</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/STAR/STARsolo/Gene/raw/matrix.mtx.gz&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">colnames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/STAR/STARsolo/Gene/raw/barcodes.tsv.gz&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">rownames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/STAR/STARsolo/Gene/raw/features.tsv.gz&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove empty cells</span>
</span></span><span style="display:flex;"><span>empty_cells <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(temp)
</span></span><span style="display:flex;"><span>empty_cells <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(empty_cells <span style="color:#000;font-weight:bold">==</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[,<span style="color:#000;font-weight:bold">-</span>empty_cells]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_all_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h5 id="filtered-cellular-barcodes-1">Filtered cellular barcodes</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">readMM</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/STAR/STARsolo/Gene/filtered/matrix.mtx.gz&#34;</span>)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">colnames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/STAR/STARsolo/Gene/filtered/barcodes.tsv.gz&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">rownames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/STAR/STARsolo/Gene/filtered/features.tsv.gz&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_filtered_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h5 id="cellular-barcodes-from-cell-ranger">Cellular barcodes from Cell Ranger</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>cells <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">colnames</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">gsub</span>(pattern <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;-1&#39;</span>, replacement <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_all_barcodes[,cells]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h4 id="alevin">Alevin</h4>
<h5 id="all-cellular-barcodes-2">All cellular barcodes</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">readMM</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/alevin/raw/alevin/quants_mat.mtx.gz&#34;</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">t</span>()
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">colnames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/alevin/raw/alevin/quants_mat_rows.txt&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">rownames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/alevin/raw/alevin/quants_mat_cols.txt&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove empty cells</span>
</span></span><span style="display:flex;"><span>empty_cells <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(temp)
</span></span><span style="display:flex;"><span>empty_cells <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(empty_cells <span style="color:#000;font-weight:bold">==</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># no empty cells</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_all_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h5 id="filtered-cellular-barcodes-2">Filtered cellular barcodes</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">readMM</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/alevin/filtered/alevin/quants_mat.mtx.gz&#34;</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">t</span>()
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">colnames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/alevin/filtered/alevin/quants_mat_rows.txt&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">rownames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/alevin/filtered/alevin/quants_mat_cols.txt&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_filtered_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h5 id="cellular-barcodes-from-cell-ranger-1">Cellular barcodes from Cell Ranger</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>cells <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">colnames</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">gsub</span>(pattern <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;-1&#39;</span>, replacement <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_all_barcodes[,cells]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h4 id="kallisto-3">kallisto</h4>
<h5 id="all-cellular-barcodes-3">All cellular barcodes</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">readMM</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/kallisto/genecount/genes.mtx&#34;</span>) <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">t</span>()
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">colnames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/kallisto/genecount/genes.barcodes.txt&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">rownames</span>(temp) <span style="color:#000;font-weight:bold">&lt;-</span> readr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">read_tsv</span>(<span style="color:#d14">&#34;results/pbmc_10k_v3/kallisto/genecount/genes.genes.txt&#34;</span>, col_names <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">select</span>(<span style="color:#099">1</span>) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">t</span>() <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">as.vector</span>()
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># remove empty cells</span>
</span></span><span style="display:flex;"><span>empty_cells <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(temp)
</span></span><span style="display:flex;"><span>empty_cells <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(empty_cells <span style="color:#000;font-weight:bold">==</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[,<span style="color:#000;font-weight:bold">-</span>empty_cells]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_all_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h5 id="cellular-barcodes-from-cell-ranger-2">Cellular barcodes from Cell Ranger</h5>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>cells <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">colnames</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">gsub</span>(pattern <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;-1&#39;</span>, replacement <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_all_barcodes[,cells]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">any</span>(<span style="color:#900;font-weight:bold">duplicated</span>(<span style="color:#900;font-weight:bold">rownames</span>(temp)))
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic"># FALSE</span>
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#998;font-style:italic">## remove non-expressed genes</span>
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>(temp)
</span></span><span style="display:flex;"><span>expressed_genes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">which</span>(expressed_genes <span style="color:#000;font-weight:bold">&gt;</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp[expressed_genes,]
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> temp
</span></span></code></pre></div><h4 id="save-data">Save data</h4>
<p>Finally, we save the object that contains all the transcript count matrices.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#900;font-weight:bold">saveRDS</span>(feature_matrix, <span style="color:#d14">&#39;results/pbmc_10k_v3/merged_data/feature_matrix.rds&#39;</span>)
</span></span></code></pre></div><h3 id="color-code">Color code</h3>
<p>A coherent color code helps to quickly read the data.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>color_assignments <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">c</span>(
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Cell Ranger&#39;</span>                <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#c0392b&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;STARsolo&#39;</span>                   <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#2980b9&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Alevin (all barcodes)&#39;</span>      <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#f39c12&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Alevin (filtered barcodes)&#39;</span> <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#f39c12&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;kallisto&#39;</span>                   <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#20bf6b&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Cell Ranger (all barcodes)&#39;</span>       <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#c0392b&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Cell Ranger (filtered barcodes)&#39;</span>  <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#c0392b&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;STARsolo (all barcodes)&#39;</span>          <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#2980b9&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;STARsolo (filtered barcodes)&#39;</span>     <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#2980b9&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;STARsolo (Cell Ranger whitelist)&#39;</span> <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#2980b9&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Alevin (all barcodes)&#39;</span>            <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#f39c12&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Alevin (filtered barcodes)&#39;</span>       <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#f39c12&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Alevin (Cell Ranger whitelist)&#39;</span>   <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#f39c12&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;kallisto (all barcodes)&#39;</span>          <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#20bf6b&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;kallisto (Cell Ranger whitelist)&#39;</span> <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;#20bf6b&#39;</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div><h3 id="collect-meta-data">Collect meta data</h3>
<p>Now, we collect some meta data.</p>
<h4 id="runs">Runs</h4>
<p>Here, we put together the run statistics, including the run time, peak memory usage, etc.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#900;font-weight:bold">library</span>(lubridate)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>meta_data_runs <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tribble</span>(
</span></span><span style="display:flex;"><span>                            <span style="color:#000;font-weight:bold">~</span>run,  <span style="color:#000;font-weight:bold">~</span>run_time,  <span style="color:#000;font-weight:bold">~</span>CPU_hours,   <span style="color:#000;font-weight:bold">~</span>mem_KB,  <span style="color:#000;font-weight:bold">~</span>vmem_KB, 
</span></span><span style="display:flex;"><span>    <span style="color:#d14">&#39;Cell Ranger&#39;</span>               , <span style="color:#d14">&#39;20:39:07&#39;</span>, <span style="color:#d14">&#39;222:30:36&#39;</span>, <span style="color:#099">125829116</span>, <span style="color:#099">125829120</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#d14">&#39;STARsolo&#39;</span>                  , <span style="color:#d14">&#39;00:34:45&#39;</span>,  <span style="color:#d14">&#39;09:10:46&#39;</span>,  <span style="color:#099">45942620</span>,  <span style="color:#099">46475712</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#d14">&#39;Alevin (all barcodes)&#39;</span>     , <span style="color:#d14">&#39;03:09:25&#39;</span>,  <span style="color:#d14">&#39;53:05:08&#39;</span>,  <span style="color:#099">21223076</span>,  <span style="color:#099">34565292</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#d14">&#39;Alevin (filtered barcodes)&#39;</span>, <span style="color:#d14">&#39;00:59:45&#39;</span>,  <span style="color:#d14">&#39;13:06:48&#39;</span>,  <span style="color:#099">18361588</span>,  <span style="color:#099">33971372</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#d14">&#39;kallisto&#39;</span>                  , <span style="color:#d14">&#39;00:35:09&#39;</span>,  <span style="color:#d14">&#39;01:54:21&#39;</span>,  <span style="color:#099">13737072</span>,  <span style="color:#099">14877188</span>
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(
</span></span><span style="display:flex;"><span>    run <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(run, levels <span style="color:#000;font-weight:bold">=</span> run),
</span></span><span style="display:flex;"><span>    run_time <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">hms</span>(run_time),
</span></span><span style="display:flex;"><span>    run_time <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">hour</span>(run_time)<span style="color:#000;font-weight:bold">*</span><span style="color:#099">60</span> <span style="color:#000;font-weight:bold">+</span> <span style="color:#900;font-weight:bold">minute</span>(run_time) <span style="color:#000;font-weight:bold">+</span> <span style="color:#900;font-weight:bold">second</span>(run_time)<span style="color:#000;font-weight:bold">/</span><span style="color:#099">60</span>,
</span></span><span style="display:flex;"><span>    CPU_hours <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">hms</span>(CPU_hours),
</span></span><span style="display:flex;"><span>    CPU_hours <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">hour</span>(CPU_hours) <span style="color:#000;font-weight:bold">+</span> <span style="color:#900;font-weight:bold">minute</span>(CPU_hours)<span style="color:#000;font-weight:bold">/</span><span style="color:#099">60</span>,
</span></span><span style="display:flex;"><span>    mem_GB <span style="color:#000;font-weight:bold">=</span> mem_KB<span style="color:#000;font-weight:bold">/</span><span style="color:#099">1E6</span>,
</span></span><span style="display:flex;"><span>    vmem_GB <span style="color:#000;font-weight:bold">=</span> vmem_KB<span style="color:#000;font-weight:bold">/</span><span style="color:#099">1E6</span>
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">saveRDS</span>(meta_data_runs, <span style="color:#d14">&#39;results/pbmc_10k_v3/merged_data/meta_data_runs.rds&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>knitr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">kable</span>(meta_data_runs)
</span></span></code></pre></div><table>
<thead>
<tr>
<th style="text-align:left"></th>
<th style="text-align:right">Run time [min]</th>
<th style="text-align:right">CPU time [h]</th>
<th style="text-align:right">mem [GB]</th>
<th style="text-align:right">vmem [GB]</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align:left">Cell Ranger</td>
<td style="text-align:right">1239.1</td>
<td style="text-align:right">222.5</td>
<td style="text-align:right">125.8</td>
<td style="text-align:right">125.8</td>
</tr>
<tr>
<td style="text-align:left">STARsolo</td>
<td style="text-align:right">34.8</td>
<td style="text-align:right">9.2</td>
<td style="text-align:right">45.9</td>
<td style="text-align:right">46.5</td>
</tr>
<tr>
<td style="text-align:left">Alevin (all barcodes)</td>
<td style="text-align:right">189.4</td>
<td style="text-align:right">53.1</td>
<td style="text-align:right">21.2</td>
<td style="text-align:right">34.6</td>
</tr>
<tr>
<td style="text-align:left">Alevin (filtered barcodes)</td>
<td style="text-align:right">59.8</td>
<td style="text-align:right">13.1</td>
<td style="text-align:right">18.4</td>
<td style="text-align:right">34.0</td>
</tr>
<tr>
<td style="text-align:left">kallisto</td>
<td style="text-align:right">35.2</td>
<td style="text-align:right">1.9</td>
<td style="text-align:right">13.7</td>
<td style="text-align:right">14.9</td>
</tr>
</tbody>
</table>
<h4 id="cells">Cells</h4>
<p>Let&rsquo;s define the group levels.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>group_levels <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">c</span>(
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Cell Ranger (all barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Cell Ranger (filtered barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;STARsolo (all barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;STARsolo (filtered barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;STARsolo (Cell Ranger whitelist)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Alevin (all barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Alevin (filtered barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;Alevin (Cell Ranger whitelist)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;kallisto (all barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;kallisto (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div><p>Here, for every transcript count matrix we got, we count the number of transcripts and number of expressed genes per cell and rank them by the number of transcripts (we will need that later).</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp_cell_ranger_all_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Cell Ranger (all barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_all_barcodes <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Cell Ranger (filtered barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_starsolo_all_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;STARsolo (all barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_all_barcodes <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_starsolo_filtered_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;STARsolo (filtered barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_filtered_barcodes <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;STARsolo (Cell Ranger whitelist)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_alevin_all_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Alevin (all barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_all_barcodes <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_alevin_filtered_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Alevin (filtered barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_filtered_barcodes <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Alevin (Cell Ranger whitelist)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_kallisto_all_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;kallisto (all barcodes)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_all_barcodes <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;kallisto (Cell Ranger whitelist)&#39;</span>,
</span></span><span style="display:flex;"><span>    cell <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">colnames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(),
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">colSums</span>(feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">!=</span> <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">arrange</span>(<span style="color:#000;font-weight:bold">-</span>nUMI) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(rank <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">1</span><span style="color:#000;font-weight:bold">:</span><span style="color:#900;font-weight:bold">nrow</span>(.)))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>meta_data_cells <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">bind_rows</span>(
</span></span><span style="display:flex;"><span>    temp_cell_ranger_all_barcodes,
</span></span><span style="display:flex;"><span>    temp_cell_ranger_filtered_barcodes,
</span></span><span style="display:flex;"><span>    temp_starsolo_all_barcodes,
</span></span><span style="display:flex;"><span>    temp_starsolo_filtered_barcodes,
</span></span><span style="display:flex;"><span>    temp_starsolo_cell_ranger_whitelist,
</span></span><span style="display:flex;"><span>    temp_alevin_all_barcodes,
</span></span><span style="display:flex;"><span>    temp_alevin_filtered_barcodes,
</span></span><span style="display:flex;"><span>    temp_alevin_cell_ranger_whitelist,
</span></span><span style="display:flex;"><span>    temp_kallisto_all_barcodes,
</span></span><span style="display:flex;"><span>    temp_kallisto_cell_ranger_whitelist
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(group <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(group, levels <span style="color:#000;font-weight:bold">=</span> group_levels))
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">saveRDS</span>(meta_data_cells, <span style="color:#d14">&#39;results/pbmc_10k_v3/merged_data/meta_data_cells.rds&#39;</span>)
</span></span></code></pre></div><h3 id="run-time">Run time</h3>
<p>The run time is defined as the time it took a job to finish after it was started.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_runs <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(run <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">fct_rev</span>(run)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(run_time, run, fill <span style="color:#000;font-weight:bold">=</span> run)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_bar</span>(stat <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;identity&#39;</span>, position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;dodge&#39;</span>, color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;black&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_label</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(
</span></span><span style="display:flex;"><span>      hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">ifelse</span>(run <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span>, <span style="color:#099">1.2</span>, <span style="color:#099">-0.2</span>),
</span></span><span style="display:flex;"><span>      label <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">comma</span>(<span style="color:#900;font-weight:bold">round</span>(run_time, digits <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>))
</span></span><span style="display:flex;"><span>    ),
</span></span><span style="display:flex;"><span>    fill <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;white&#39;</span>, alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.75</span>
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_continuous</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Run time [minutes]&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma, expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.01</span>, <span style="color:#099">0</span>),
</span></span><span style="display:flex;"><span>    breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">0</span>, <span style="color:#099">1500</span>, by <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">200</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_discrete</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_manual</span>(values <span style="color:#000;font-weight:bold">=</span> color_assignments) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.major.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/run_time.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">2.5</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">8</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/run_time.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/run_time.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/run_time.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p>STARsolo and kallisto both finished in just 35 minutes, while Alevin needed either about 60 minutes with a filtered whitelist or almost 200 minutes when quantifying the transcript counts for all valid cellular barcodes.
Cell Ranger, which internally uses STAR but not the STARsolo mode, needed more than 20 hours.</p>
<h3 id="cpu-hours">CPU hours</h3>
<p>The measure &ldquo;CPU hours&rdquo; reflects exactly what it says: the cumulative hours used across all the cores that were engaged in the job.
For example, a 30 minute long job using 10 cores would have the same CPU hours as a 1 hour long job using 5 cores (assuming that all cores were engaged the entire time): 5 CPU hours.
This measure is taken from the run summary of the PBE scheduler and I cannot comment on how reliable it is.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_runs <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(run <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">fct_rev</span>(run)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(CPU_hours, run, fill <span style="color:#000;font-weight:bold">=</span> run)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_bar</span>(stat <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;identity&#39;</span>, position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;dodge&#39;</span>, color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;black&#39;</span>, show.legend <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">FALSE</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_label</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(
</span></span><span style="display:flex;"><span>      hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">ifelse</span>(run <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span>, <span style="color:#099">1.2</span>, <span style="color:#099">-0.2</span>),
</span></span><span style="display:flex;"><span>      label <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">comma</span>(<span style="color:#900;font-weight:bold">round</span>(CPU_hours, digits <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>))
</span></span><span style="display:flex;"><span>    ),
</span></span><span style="display:flex;"><span>    fill <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;white&#39;</span>, alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.75</span>
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_continuous</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;CPU time [hours]&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma, expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.01</span>, <span style="color:#099">0</span>),
</span></span><span style="display:flex;"><span>    breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">0</span>, <span style="color:#099">300</span>, by <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">50</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_discrete</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_manual</span>(values <span style="color:#000;font-weight:bold">=</span> color_assignments) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.major.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/cpu_hours.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">2.5</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">8</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/cpu_hours.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/cpu_hours.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/cpu_hours.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p>As expected, since in principle every job had the same number of cores available (20), Cell Ranger accumulated the most CPU hours, with Alevin (all barcodes) trailing behind.
Interestingly, STARsolo used more CPU hours than kallisto, even though they had the same run time.
While this is impressive, it is also important to keep in mind that STAR aligns the reads to the genome, while kallisto (only) analyses the transcriptome, which of course is much smaller.</p>
<h3 id="peak-memory-usage">Peak memory usage</h3>
<p>It would be great to have a temporal profile of memory occupancy during the job, but unfortunately that appears to be quite complicated to record.
Instead, here we see the memory usage as reported by the PBE job scheduler.
As far as I know, this number corresponds to the highest (peak) usage throughout the entire job, meaning that we don&rsquo;t know anything about how this value potentially fluctuates.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_runs <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(run <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">fct_rev</span>(run)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rename</span>(mem <span style="color:#000;font-weight:bold">=</span> mem_GB, vmem <span style="color:#000;font-weight:bold">=</span> vmem_GB) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">pivot_longer</span>(
</span></span><span style="display:flex;"><span>   cols <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#d14">&#39;mem&#39;</span>,<span style="color:#d14">&#39;vmem&#39;</span>),
</span></span><span style="display:flex;"><span>   names_to <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;category&#39;</span>,
</span></span><span style="display:flex;"><span>   values_to <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;memory&#39;</span>
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(category <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(category, levels <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#d14">&#39;vmem&#39;</span>,<span style="color:#d14">&#39;mem&#39;</span>))) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(memory, run, fill <span style="color:#000;font-weight:bold">=</span> category)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_col</span>(position <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">position_dodge</span>(width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>), color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;black&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_label</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(
</span></span><span style="display:flex;"><span>      group <span style="color:#000;font-weight:bold">=</span> category, label <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">format</span>(<span style="color:#900;font-weight:bold">round</span>(memory, digits <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>), trim <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">TRUE</span>),
</span></span><span style="display:flex;"><span>      hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">ifelse</span>(run <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span>, <span style="color:#099">1.2</span>, <span style="color:#099">-0.2</span>)
</span></span><span style="display:flex;"><span>    ),
</span></span><span style="display:flex;"><span>    position <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">position_dodge</span>(width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>), fill <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;white&#39;</span>, alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.75</span>,
</span></span><span style="display:flex;"><span>    size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3</span>, label.padding <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">unit</span>(<span style="color:#099">0.1</span>, <span style="color:#d14">&#39;lines&#39;</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_continuous</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Memory [GB]&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma, expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.01</span>, <span style="color:#099">0</span>)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_discrete</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_brewer</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>, palette <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Set1&#39;</span>, direction <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">-1</span>,
</span></span><span style="display:flex;"><span>    guide <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">guide_legend</span>(reverse <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">TRUE</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;right&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.major.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/memory_footprint.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">2.5</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">8</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/memory_footprint.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/memory_footprint.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/memory_footprint.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p>Here we have some surprising results, or at least I think so.
I&rsquo;m not too familiar with the distinction between physical and virtual memory, but I don&rsquo;t know if it makes sense that they have almost identical values in every run.
Then, perhaps one would assume that, since Cell Ranger takes so long and it uses so many CPU hours, at least it doesn&rsquo;t need a lot of memory.
But apparently that is not the case, as Cell Ranger uses all of the memory it was given (at least momentarily, possibly the whole time).
Also, when analysing all valid cellular barcodes, Alevin scores the lowest memory usage, despite the high CPU hours it needed.
Again, kallisto performs well with a memory usage of less than 40 GB.</p>
<h3 id="number-of-cells">Number of cells</h3>
<p>Next, I looked at the number of cells that were identified in each run.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">group_by</span>(group) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">summarise</span>(cell_count <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">n</span>()) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(group <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">fct_rev</span>(group)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(cell_count, group, fill <span style="color:#000;font-weight:bold">=</span> group)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_col</span>(position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;dodge&#39;</span>, color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;black&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_label</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(label <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">format</span>(<span style="color:#900;font-weight:bold">round</span>(cell_count, digits <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>), big.mark <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;,&#39;</span>, trim <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">TRUE</span>)),
</span></span><span style="display:flex;"><span>    hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1.1</span>, position <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">position_dodge</span>(width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>), fill <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;white&#39;</span>, alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.75</span>,
</span></span><span style="display:flex;"><span>    size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">4</span>
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of cells&#39;</span>, expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.01</span>, <span style="color:#099">0</span>),
</span></span><span style="display:flex;"><span>    breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E0</span>, <span style="color:#099">1E1</span>,<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>,<span style="color:#099">1E6</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_discrete</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>, expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.1</span>, <span style="color:#099">0</span>)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_manual</span>(values <span style="color:#000;font-weight:bold">=</span> color_assignments) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">annotation_logticks</span>(
</span></span><span style="display:flex;"><span>    side <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;b&#39;</span>, short <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">unit</span>(<span style="color:#099">0.15</span>, <span style="color:#d14">&#39;cm&#39;</span>), mid <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">unit</span>(<span style="color:#099">0.15</span>, <span style="color:#d14">&#39;cm&#39;</span>),
</span></span><span style="display:flex;"><span>    long <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">unit</span>(<span style="color:#099">0.2</span>, <span style="color:#d14">&#39;cm&#39;</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.minor.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>(),
</span></span><span style="display:flex;"><span>    panel.grid.major.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/number_of_cells.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">4</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">8</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/number_of_cells.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/number_of_cells.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/number_of_cells.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p>To no suprise, when providing the whitelist generated by Cell Ranger, Alevin extracted the same cellular barcodes.
Conveniently, both STARsolo and kallisto only needed to be run once and provide us with the transcript counts for all cellular barcodes, meaning that we can simply extract the cells identified by Cell Ranger.
STARsolo and Alevin, using their own filtering algorithms, retrieved similar cell counts as Cell Ranger.
No drastic differences are seen in the total barcode counts, which are about 1.1 million (Alevin is slightly below the other three).</p>
<h3 id="total-number-of-transcriptsmolecules">Total number of transcripts/molecules</h3>
<p>Next, I looked at the number of transcripts that were counted across all cells and genes.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">group_by</span>(group) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">summarise</span>(nUMI <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">sum</span>(nUMI)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(group <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">fct_rev</span>(group)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(nUMI, group, fill <span style="color:#000;font-weight:bold">=</span> group)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_col</span>(color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;black&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_label</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(label <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">comma</span>(nUMI)), hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1.1</span>, fill <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;white&#39;</span>,
</span></span><span style="display:flex;"><span>    alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.75</span>
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_continuous</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Total number of transcripts&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma,
</span></span><span style="display:flex;"><span>    expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.01</span>, <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_discrete</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_manual</span>(values <span style="color:#000;font-weight:bold">=</span> color_assignments) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.minor.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>(),
</span></span><span style="display:flex;"><span>    panel.grid.major.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/number_of_transcripts.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">4</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">8</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/number_of_transcripts.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/number_of_transcripts.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/number_of_transcripts.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p>In all cases, roughly 10% of all transcripts (across all cellular barcodes) belong to barcodes that are later filtered out.
When using the Cell Ranger whitelist, very similar counts are seen for Cell Ranger, STARsolo, and Alevin.
kallisto counted considerably (10%) more transcripts, both for all and filtered barcodes.</p>
<h3 id="number-of-expressed-genes">Number of expressed genes</h3>
<p>Similar to the previous analysis, I also checked how many genes are expressed in at least 1 cell with at least 1 transcript.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    group <span style="color:#000;font-weight:bold">=</span> group_levels,
</span></span><span style="display:flex;"><span>    nGene <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_all_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>(),
</span></span><span style="display:flex;"><span>      feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">nrow</span>()
</span></span><span style="display:flex;"><span>    )
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  dplyr<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">mutate</span>(group <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(group, levels <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">rev</span>(group_levels))) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(nGene, group, fill <span style="color:#000;font-weight:bold">=</span> group)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_col</span>(color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;black&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_label</span>(
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(label <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">comma</span>(nGene)), hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1.1</span>, fill <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;white&#39;</span>,
</span></span><span style="display:flex;"><span>    alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.75</span>
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_continuous</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of expressed genes&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma,
</span></span><span style="display:flex;"><span>    , expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.01</span>, <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_discrete</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_manual</span>(values <span style="color:#000;font-weight:bold">=</span> color_assignments) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.minor.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>(),
</span></span><span style="display:flex;"><span>    panel.grid.major.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/number_of_expressed_genes.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">4</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">8</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/number_of_expressed_genes.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/number_of_expressed_genes.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/number_of_expressed_genes.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p>As expected, slightly more genes are expressed in the matrices containing all cellular barcodes, compared to those containing only the filtered cells.
While Cell Ranger, STARsolo, and Alevin again provide similar numbers, kallisto found about 20% more genes to be expressed.</p>
<h3 id="correlation-of-cells">Correlation of cells</h3>
<p>Here, we focus on those cells whose expression profiles were quantified with all tools, namely those in the Cell Ranger whitelist.</p>
<h4 id="number-of-transcripts">Number of transcripts</h4>
<p>First, I compared the number of transcripts that were identified in each cell for every comparison of tools.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span><span style="color:#900;font-weight:bold">library</span>(patchwork)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">library</span>(viridis)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_groups <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#d14">&#39;Cell Ranger&#39;</span>, <span style="color:#d14">&#39;STARsolo&#39;</span>, <span style="color:#d14">&#39;Alevin&#39;</span>, <span style="color:#d14">&#39;kallisto&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>  group.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">character</span>(),
</span></span><span style="display:flex;"><span>  group.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">character</span>(),
</span></span><span style="display:flex;"><span>  nUMI.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>(),
</span></span><span style="display:flex;"><span>  nUMI.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>()
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">for </span>( i in temp_groups ) {
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">for </span>( j in temp_groups ) {
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">## assign correct name to group 1</span>
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;Cell Ranger (filtered barcodes)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;STARsolo&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;STARsolo (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Alevin&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;Alevin (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;kallisto&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;kallisto (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">## assign correct name to group 2</span>
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;Cell Ranger (filtered barcodes)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;STARsolo&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;STARsolo (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Alevin&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;Alevin (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;kallisto&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;kallisto (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">## </span>
</span></span><span style="display:flex;"><span>    temp <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">bind_rows</span>(
</span></span><span style="display:flex;"><span>      temp,
</span></span><span style="display:flex;"><span>      <span style="color:#900;font-weight:bold">full_join</span>(
</span></span><span style="display:flex;"><span>        meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">filter</span>(group <span style="color:#000;font-weight:bold">==</span> temp_group_1) <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">select</span>(cell, nUMI),
</span></span><span style="display:flex;"><span>        meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">filter</span>(group <span style="color:#000;font-weight:bold">==</span> temp_group_2) <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">select</span>(cell, nUMI),
</span></span><span style="display:flex;"><span>        by <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;cell&#39;</span>
</span></span><span style="display:flex;"><span>      ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>      <span style="color:#900;font-weight:bold">mutate</span>(group.x <span style="color:#000;font-weight:bold">=</span> i, group.y <span style="color:#000;font-weight:bold">=</span> j)
</span></span><span style="display:flex;"><span>    )
</span></span><span style="display:flex;"><span>  }
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(
</span></span><span style="display:flex;"><span>    group.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(group.x, levels <span style="color:#000;font-weight:bold">=</span> temp_groups),
</span></span><span style="display:flex;"><span>    group.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(group.y, levels <span style="color:#000;font-weight:bold">=</span> temp_groups)
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_labels <span style="color:#000;font-weight:bold">&lt;-</span> temp <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">group_by</span>(group.x, group.y) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">summarise</span>(n <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">n</span>())
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">ggplot</span>(temp, <span style="color:#900;font-weight:bold">aes</span>(nUMI.x, nUMI.y)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_point</span>(size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.2</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_abline</span>(intercept <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, slope <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>, color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;red&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_text</span>(
</span></span><span style="display:flex;"><span>    data <span style="color:#000;font-weight:bold">=</span> temp_labels, hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(x <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">5E2</span>, y <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">7E4</span>, label <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">paste0</span>(<span style="color:#d14">&#39;n = &#39;</span>, <span style="color:#900;font-weight:bold">format</span>(n, big.mark <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;,&#39;</span>)))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>,<span style="color:#099">1E6</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>,<span style="color:#099">1E6</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">annotation_logticks</span>(side <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bl&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">coord_fixed</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    panel.grid.minor <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>(),
</span></span><span style="display:flex;"><span>    strip.text <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_text</span>(face <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bold&#39;</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">facet_grid</span>(group.y <span style="color:#000;font-weight:bold">~</span> group.x)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;results/pbmc_10k_v3/plots/correlation_cells_number_of_transcripts_1.png&#39;</span>,
</span></span><span style="display:flex;"><span>  p,
</span></span><span style="display:flex;"><span>  height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">ggplot</span>(temp, <span style="color:#900;font-weight:bold">aes</span>(nUMI.x, nUMI.y)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">stat_density_2d</span>(<span style="color:#900;font-weight:bold">aes</span>(fill <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">stat</span>(level)), geom <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;polygon&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_abline</span>(intercept <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, slope <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>, color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;red&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>,<span style="color:#099">1E6</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>,<span style="color:#099">1E6</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_viridis</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">annotation_logticks</span>(side <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bl&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">coord_fixed</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.minor <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>(),
</span></span><span style="display:flex;"><span>    strip.text <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_text</span>(face <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bold&#39;</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">facet_grid</span>(group.y <span style="color:#000;font-weight:bold">~</span> group.x)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;results/pbmc_10k_v3/plots/correlation_cells_number_of_transcripts_2.png&#39;</span>,
</span></span><span style="display:flex;"><span>  p,
</span></span><span style="display:flex;"><span>  height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/correlation_cells_number_of_transcripts_1.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/correlation_cells_number_of_transcripts_1.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/correlation_cells_number_of_transcripts_1.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p>In most comparisons, the cells are distributed closely along the diagonal line (perfect correlation).
However, consistent with the higher number of transcripts identified by kallisto, the majority of cells are located above the diagonal line, meaning that kallisto identified more transcripts evenly across all cells.</p>
<h4 id="number-of-expressed-genes-1">Number of expressed genes</h4>
<p>Next, we checked the number of expressed genes (at least 1 transcript) across the cells.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp_groups <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#d14">&#39;Cell Ranger&#39;</span>, <span style="color:#d14">&#39;STARsolo&#39;</span>, <span style="color:#d14">&#39;Alevin&#39;</span>, <span style="color:#d14">&#39;kallisto&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>  group.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">character</span>(),
</span></span><span style="display:flex;"><span>  group.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">character</span>(),
</span></span><span style="display:flex;"><span>  nGene.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>(),
</span></span><span style="display:flex;"><span>  nGene.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>()
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">for </span>( i in temp_groups ) {
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">for </span>( j in temp_groups ) {
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">## assign correct name to group 1</span>
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;Cell Ranger (filtered barcodes)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;STARsolo&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;STARsolo (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Alevin&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;Alevin (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;kallisto&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;kallisto (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">## assign correct name to group 2</span>
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;Cell Ranger (filtered barcodes)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;STARsolo&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;STARsolo (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Alevin&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;Alevin (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;kallisto&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#d14">&#39;kallisto (Cell Ranger whitelist)&#39;</span>
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">## </span>
</span></span><span style="display:flex;"><span>    temp <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">bind_rows</span>(
</span></span><span style="display:flex;"><span>      temp,
</span></span><span style="display:flex;"><span>      <span style="color:#900;font-weight:bold">full_join</span>(
</span></span><span style="display:flex;"><span>        meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">filter</span>(group <span style="color:#000;font-weight:bold">==</span> temp_group_1) <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">select</span>(cell, nGene),
</span></span><span style="display:flex;"><span>        meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">filter</span>(group <span style="color:#000;font-weight:bold">==</span> temp_group_2) <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">select</span>(cell, nGene),
</span></span><span style="display:flex;"><span>        by <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;cell&#39;</span>
</span></span><span style="display:flex;"><span>      ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>      <span style="color:#900;font-weight:bold">mutate</span>(group.x <span style="color:#000;font-weight:bold">=</span> i, group.y <span style="color:#000;font-weight:bold">=</span> j)
</span></span><span style="display:flex;"><span>    )
</span></span><span style="display:flex;"><span>  }
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(
</span></span><span style="display:flex;"><span>    group.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(group.x, levels <span style="color:#000;font-weight:bold">=</span> temp_groups),
</span></span><span style="display:flex;"><span>    group.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(group.y, levels <span style="color:#000;font-weight:bold">=</span> temp_groups)
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_labels <span style="color:#000;font-weight:bold">&lt;-</span> temp <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">group_by</span>(group.x, group.y) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">summarise</span>(n <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">n</span>())
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">ggplot</span>(temp, <span style="color:#900;font-weight:bold">aes</span>(nGene.x, nGene.y)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_point</span>(size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.2</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_abline</span>(intercept <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, slope <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>, color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;red&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_text</span>(
</span></span><span style="display:flex;"><span>    data <span style="color:#000;font-weight:bold">=</span> temp_labels, hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(x <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">9E1</span>, y <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">7E3</span>, label <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">paste0</span>(<span style="color:#d14">&#39;n = &#39;</span>, <span style="color:#900;font-weight:bold">format</span>(n, big.mark <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;,&#39;</span>)))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of expressed genes&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of expressed genes&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">annotation_logticks</span>(side <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bl&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">coord_fixed</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    panel.grid.minor <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>(),
</span></span><span style="display:flex;"><span>    strip.text <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_text</span>(face <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bold&#39;</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">facet_grid</span>(group.y <span style="color:#000;font-weight:bold">~</span> group.x)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;results/pbmc_10k_v3/plots/correlation_cells_number_of_expressed_genes.png&#39;</span>,
</span></span><span style="display:flex;"><span>  p,
</span></span><span style="display:flex;"><span>  height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/correlation_cells_number_of_expressed_genes.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/correlation_cells_number_of_expressed_genes.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/correlation_cells_number_of_expressed_genes.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p>While Cell Ranger, STAR, and Alevin have comparable results, kallisto found more genes to be expressed in most cells.</p>
<h3 id="correlation-of-transcripts-per-gene">Correlation of transcripts per gene</h3>
<p>Here, we switch focus to genes instead of cells and ask how many transcripts were found for every gene between the tools.</p>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp_groups <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#d14">&#39;Cell Ranger&#39;</span>, <span style="color:#d14">&#39;STARsolo&#39;</span>, <span style="color:#d14">&#39;Alevin&#39;</span>, <span style="color:#d14">&#39;kallisto&#39;</span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    gene <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">rownames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>cell_ranger_filtered_barcodes <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    gene <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">rownames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>starsolo_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    gene <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">rownames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>alevin_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>    gene <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> <span style="color:#900;font-weight:bold">rownames</span>(),
</span></span><span style="display:flex;"><span>    nUMI <span style="color:#000;font-weight:bold">=</span> feature_matrix<span style="color:#000;font-weight:bold">$</span>kallisto_cell_ranger_whitelist <span style="color:#000;font-weight:bold">%&gt;%</span> Matrix<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">rowSums</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">tibble</span>(
</span></span><span style="display:flex;"><span>  group.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">character</span>(),
</span></span><span style="display:flex;"><span>  group.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">character</span>(),
</span></span><span style="display:flex;"><span>  nUMI.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>(),
</span></span><span style="display:flex;"><span>  nUMI.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">numeric</span>()
</span></span><span style="display:flex;"><span>)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">for </span>( i in temp_groups ) {
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">for </span>( j in temp_groups ) {
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">## assign correct name to group 1</span>
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> temp_cell_ranger_filtered_barcodes
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;STARsolo&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> temp_starsolo_cell_ranger_whitelist
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Alevin&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> temp_alevin_cell_ranger_whitelist
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( i <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;kallisto&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_1 <span style="color:#000;font-weight:bold">&lt;-</span> temp_kallisto_cell_ranger_whitelist
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">## assign correct name to group 2</span>
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Cell Ranger&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> temp_cell_ranger_filtered_barcodes
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;STARsolo&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> temp_starsolo_cell_ranger_whitelist
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;Alevin&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> temp_alevin_cell_ranger_whitelist
</span></span><span style="display:flex;"><span>    } else <span style="color:#900;font-weight:bold">if </span>( j <span style="color:#000;font-weight:bold">==</span> <span style="color:#d14">&#39;kallisto&#39;</span> ) {
</span></span><span style="display:flex;"><span>      temp_group_2 <span style="color:#000;font-weight:bold">&lt;-</span> temp_kallisto_cell_ranger_whitelist
</span></span><span style="display:flex;"><span>    }
</span></span><span style="display:flex;"><span>    <span style="color:#998;font-style:italic">##</span>
</span></span><span style="display:flex;"><span>    temp <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">bind_rows</span>(
</span></span><span style="display:flex;"><span>      temp,
</span></span><span style="display:flex;"><span>      <span style="color:#900;font-weight:bold">full_join</span>(
</span></span><span style="display:flex;"><span>        temp_group_1,
</span></span><span style="display:flex;"><span>        temp_group_2,
</span></span><span style="display:flex;"><span>        by <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;gene&#39;</span>
</span></span><span style="display:flex;"><span>      ) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>      <span style="color:#900;font-weight:bold">replace_na</span>(<span style="color:#900;font-weight:bold">list</span>(nUMI.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.1</span>, nUMI.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.1</span>)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>      <span style="color:#900;font-weight:bold">mutate</span>(
</span></span><span style="display:flex;"><span>        nUMI.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">ifelse</span>(nUMI.x <span style="color:#000;font-weight:bold">&lt;</span> <span style="color:#099">0.1</span>, <span style="color:#099">0.1</span>, nUMI.x),
</span></span><span style="display:flex;"><span>        nUMI.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">ifelse</span>(nUMI.y <span style="color:#000;font-weight:bold">&lt;</span> <span style="color:#099">0.1</span>, <span style="color:#099">0.1</span>, nUMI.y),
</span></span><span style="display:flex;"><span>        group.x <span style="color:#000;font-weight:bold">=</span> i, group.y <span style="color:#000;font-weight:bold">=</span> j
</span></span><span style="display:flex;"><span>      )
</span></span><span style="display:flex;"><span>    )
</span></span><span style="display:flex;"><span>  }
</span></span><span style="display:flex;"><span>}
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp <span style="color:#000;font-weight:bold">&lt;-</span> temp <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(
</span></span><span style="display:flex;"><span>    group.x <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(group.x, levels <span style="color:#000;font-weight:bold">=</span> temp_groups),
</span></span><span style="display:flex;"><span>    group.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">factor</span>(group.y, levels <span style="color:#000;font-weight:bold">=</span> temp_groups)
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>temp_labels <span style="color:#000;font-weight:bold">&lt;-</span> temp <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">group_by</span>(group.x, group.y) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">summarise</span>(n <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">n</span>())
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> <span style="color:#900;font-weight:bold">ggplot</span>(temp, <span style="color:#900;font-weight:bold">aes</span>(nUMI.x, nUMI.y)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_point</span>(size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.5</span>, alpha <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.5</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_abline</span>(intercept <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, slope <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">1</span>, color <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;red&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_text</span>(
</span></span><span style="display:flex;"><span>    data <span style="color:#000;font-weight:bold">=</span> temp_labels, hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(x <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">5E-1</span>, y <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3E6</span>, label <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">paste0</span>(<span style="color:#d14">&#39;n = &#39;</span>, <span style="color:#900;font-weight:bold">format</span>(n, big.mark <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;,&#39;</span>)))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E-1</span>,<span style="color:#099">1E0</span>,<span style="color:#099">1E1</span>,<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>,<span style="color:#099">1E6</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E-1</span>,<span style="color:#099">1E0</span>,<span style="color:#099">1E1</span>,<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>,<span style="color:#099">1E6</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">annotation_logticks</span>(side <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bl&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">coord_fixed</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    panel.grid.minor <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>(),
</span></span><span style="display:flex;"><span>    strip.text <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_text</span>(face <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bold&#39;</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">facet_grid</span>(group.y <span style="color:#000;font-weight:bold">~</span> group.x)
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(
</span></span><span style="display:flex;"><span>  <span style="color:#d14">&#39;results/pbmc_10k_v3/plots/correlation_genes_number_of_transcripts.png&#39;</span>,
</span></span><span style="display:flex;"><span>  p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>
</span></span><span style="display:flex;"><span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/correlation_genes_number_of_transcripts.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/correlation_genes_number_of_transcripts.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/correlation_genes_number_of_transcripts.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<p><strong>NOTE</strong>: Genes with zero counts or counts lower than <code>0.1</code> were manually set to <code>0.1</code> to fit them in the scale.</p>
<p>The results here are more diverse than for the cells.
While the scatter plot might indicate that many genes have drastically different transcript counts between the tools, the majority is actually located along the diagonal.
Interestingly, despite the higher number of genes detected by kallisto compared to the other tools, some genes were identified in the other tools but not in kallisto.
While genes with very high counts are more consistently quantified between the tools, some genes show large discrepancies, with zero counts in one tools and more than 10,000 in another.</p>
<h3 id="knee-plots">Knee plots</h3>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp_labels <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">group_by</span>(group) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">summarise</span>(n <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">n</span>())
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(rank, nUMI)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_point</span>(size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.5</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_text</span>(
</span></span><span style="display:flex;"><span>    data <span style="color:#000;font-weight:bold">=</span> temp_labels, hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(x <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3E0</span>, y <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3E0</span>, label <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">paste0</span>(<span style="color:#d14">&#39;n = &#39;</span>, <span style="color:#900;font-weight:bold">format</span>(n, big.mark <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;,&#39;</span>, trim <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">TRUE</span>)))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Rank of cellular barcode&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E0</span>,<span style="color:#099">1E1</span>,<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>,<span style="color:#099">1E6</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_log10</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">1E0</span>,<span style="color:#099">1E1</span>,<span style="color:#099">1E2</span>,<span style="color:#099">1E3</span>,<span style="color:#099">1E4</span>,<span style="color:#099">1E5</span>),
</span></span><span style="display:flex;"><span>    labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">trans_format</span>(<span style="color:#d14">&#39;log10&#39;</span>, scales<span style="color:#000;font-weight:bold">::</span><span style="color:#900;font-weight:bold">math_format</span>(<span style="color:#099">10</span>^.x))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">annotation_logticks</span>(side <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bl&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    panel.grid.minor <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>(),
</span></span><span style="display:flex;"><span>    strip.text <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_text</span>(face <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bold&#39;</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">facet_wrap</span>(<span style="color:#000;font-weight:bold">~</span>group, ncol <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">4</span>)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/knee_plots.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">9</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">12</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/knee_plots.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/knee_plots.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/knee_plots.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<h3 id="transcripts-per-cell">Transcripts per cell</h3>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(group <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">fct_rev</span>(group)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(nUMI, group, fill <span style="color:#000;font-weight:bold">=</span> group)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_boxplot</span>(width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.5</span>, outlier.size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.2</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_continuous</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma, expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.01</span>, <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_discrete</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_manual</span>(values <span style="color:#000;font-weight:bold">=</span> color_assignments) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.major.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/distribution_transcripts_per_cell.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">4</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/distribution_transcripts_per_cell.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/distribution_transcripts_per_cell.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/distribution_transcripts_per_cell.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<h3 id="expressed-genes-per-cell">Expressed genes per cell</h3>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">mutate</span>(group <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">fct_rev</span>(group)) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(nGene, group, fill <span style="color:#000;font-weight:bold">=</span> group)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_boxplot</span>(width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.5</span>, outlier.size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.2</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_continuous</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of expressed genes&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma, expand <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">c</span>(<span style="color:#099">0.01</span>, <span style="color:#099">0</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_discrete</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;&#39;</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_fill_manual</span>(values <span style="color:#000;font-weight:bold">=</span> color_assignments) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    panel.grid.major.y <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_blank</span>()
</span></span><span style="display:flex;"><span>  )
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/distribution_expressed_genes_per_cell.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">4</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/distribution_expressed_genes_per_cell.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/distribution_expressed_genes_per_cell.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/distribution_expressed_genes_per_cell.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<h3 id="number-of-expressed-genes-over-number-of-transcripts-per-cell">Number of expressed genes over number of transcripts (per cell)</h3>
<div class="highlight"><pre tabindex="0" style="background-color:#fff;-moz-tab-size:2;-o-tab-size:2;tab-size:2;"><code class="language-r" data-lang="r"><span style="display:flex;"><span>temp_labels <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">group_by</span>(group) <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">summarise</span>(n <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">n</span>())
</span></span><span style="display:flex;"><span>
</span></span><span style="display:flex;"><span>p <span style="color:#000;font-weight:bold">&lt;-</span> meta_data_cells <span style="color:#000;font-weight:bold">%&gt;%</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">ggplot</span>(<span style="color:#900;font-weight:bold">aes</span>(nUMI, nGene)) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_point</span>(size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0.5</span>) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">geom_text</span>(
</span></span><span style="display:flex;"><span>    data <span style="color:#000;font-weight:bold">=</span> temp_labels, hjust <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">0</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">3</span>,
</span></span><span style="display:flex;"><span>    <span style="color:#900;font-weight:bold">aes</span>(x <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">52000</span>, y <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">500</span>, label <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">paste0</span>(<span style="color:#d14">&#39;n = &#39;</span>, <span style="color:#900;font-weight:bold">format</span>(n, big.mark <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;,&#39;</span>, trim <span style="color:#000;font-weight:bold">=</span> <span style="color:#000;font-weight:bold">TRUE</span>)))
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_x_continuous</span>(name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of expressed genes&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">scale_y_continuous</span>(
</span></span><span style="display:flex;"><span>    name <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;Number of transcripts&#39;</span>, labels <span style="color:#000;font-weight:bold">=</span> scales<span style="color:#000;font-weight:bold">::</span>comma,
</span></span><span style="display:flex;"><span>    breaks <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">seq</span>(<span style="color:#099">0</span>, <span style="color:#099">12000</span>, by <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">2000</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme_bw</span>() <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">theme</span>(
</span></span><span style="display:flex;"><span>    legend.position <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;none&#39;</span>,
</span></span><span style="display:flex;"><span>    strip.text <span style="color:#000;font-weight:bold">=</span> <span style="color:#900;font-weight:bold">element_text</span>(face <span style="color:#000;font-weight:bold">=</span> <span style="color:#d14">&#39;bold&#39;</span>, size <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">10</span>)
</span></span><span style="display:flex;"><span>  ) <span style="color:#000;font-weight:bold">+</span>
</span></span><span style="display:flex;"><span>  <span style="color:#900;font-weight:bold">facet_wrap</span>(<span style="color:#000;font-weight:bold">~</span>group, ncol <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">4</span>)
</span></span><span style="display:flex;"><span><span style="color:#900;font-weight:bold">ggsave</span>(<span style="color:#d14">&#39;results/pbmc_10k_v3/plots/ngene_over_numi.png&#39;</span>, p, height <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">9</span>, width <span style="color:#000;font-weight:bold">=</span> <span style="color:#099">12</span>)
</span></span></code></pre></div>
<figure
  
>
  
    <a
      href="/images/blog/benchmark_scrnaseq/ngene_over_numi.png"
      target="_blank"
    >
  
      <picture>
        <source
          srcset="/images/blog/benchmark_scrnaseq/ngene_over_numi.webp"
          type="image/webp"
        >
        <img
          src="/images/blog/benchmark_scrnaseq/ngene_over_numi.jpg"
          loading="lazy"
          
            alt="something went wrong... here should be a figure"
          
          style="display: block; margin-left: auto; margin-right: auto; width: 80%; max-width: 1000px;"
        />
      </picture>
  
    </a>
  
  
</figure>


<h2 id="to-do">To do</h2>
<ul>
<li>Add correlation coefficient to plots where necessary.</li>
<li>Try Cell Ranger 4.0.</li>
</ul>

            </div>
          </body>
        </div>
      </div>
    </div>
  </div>
  <script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
  <script src="/js/fold_code_boxes.js"></script>
  <script src="/js/line_under_header.js"></script>

      </div>
<footer class="footer">
  <div class="container">
    <div class="row">
      <div class="col-md-4 col-sm-4 text-center text-lg-left">
        &copy; Copyright Roman Hillje - All rights reserved
      </div>
      <section class="col-md-4 col-sm-4 text-center">
        <a href="https://www.twitter.com/fakechek1" target="_blank" rel="noopener noreferrer"><i class="fab fa-twitter" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a href="https://github.com/romanhaa" target="_blank" rel="noopener noreferrer"><i class="fab fa-github" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a href="https://gitlab.com/romanhaa" target="_blank" rel="noopener noreferrer"><i class="fab fa-gitlab" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a href="https://scholar.google.com/citations?hl=en&user=0yqPXREAAAAJ" target="_blank" rel="noopener noreferrer"><i class="fab fa-google" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a href="https://hub.docker.com/u/romanhaa" target="_blank" rel="noopener noreferrer"><i class="fab fa-docker" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
        <a href="https://linkedin.com/in/roman-hillje" target="_blank" rel="noopener noreferrer"><i class="fab fa-linkedin" style="color: rgba(0,0,0,.44); font-size: 2rem; margin-left: 10px" aria-hidden="true"></i></a>
      </section>
      <div class="col-md-4 col-sm-4 text-center text-lg-right">
        Adapted from <a target="_blank" href="https://www.wowthemes.net">Mediumish Theme</a>
      </div>
    </div>
  </div>
</footer>
</div>

<script src="https://code.jquery.com/jquery-3.3.1.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.3/umd/popper.min.js" integrity="sha384-ZMP7rVo3mIykV+2+9J3UJ46jBk0WLaUAdn689aCwoqbBJiSnjAK/l8WvCWPIPm49" crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/js/bootstrap.min.js" integrity="sha384-ChfqqxuZUCnJSK3+MXmPNIyE6ZbWh2IMqE241rYiqJxyMiZ6OW/JmZQ5stwEULTy" crossorigin="anonymous"></script>
<script src="/js/mediumish.js"></script>
<script src="https://cdn.rawgit.com/afeld/bootstrap-toc/v1.0.1/dist/bootstrap-toc.min.js"></script>
<script src="/js/toc.js"></script>

  </body>
</html>
